<!DOCTYPE html>
<html prefix="" lang="en">
<head>
<meta charset="utf-8">
<meta name="description" content="Science &amp; Tech for the Small Fry">
<meta name="viewport" content="width=device-width">
<title>I Love Symposia! (old posts, page 5) | I Love Symposia!</title>
<link href="assets/css/all.css" rel="stylesheet" type="text/css">
<link rel="stylesheet" href="//fonts.googleapis.com/css?family=PT+Serif:400,400italic,700%7CPT+Sans:400">
<link rel="alternate" type="application/rss+xml" title="RSS" href="rss.xml">
<link rel="canonical" href="https://ilovesymposia.com/index-5.html">
<link rel="prev" href="index-6.html" type="text/html">
<link rel="next" href="index-4.html" type="text/html">
<script type="text/x-mathjax-config">
MathJax.Hub.Config({
    tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true
    },
    displayAlign: 'center', // Change this to 'left' if you want left-aligned equations.
    "HTML-CSS": {
        styles: {'.MathJax_Display': {"margin": 0}}
    }
});
</script><!--[if lt IE 9]><script src="assets/js/html5.js"></script><![endif]-->
</head>
<body>
    <a href="#content" class="sr-only sr-only-focusable">Skip to main content</a>
    <!-- Target for toggling the sidebar `.sidebar-checkbox` is for regular
            styles, `#sidebar-checkbox` for behavior. -->
    <input type="checkbox" class="sidebar-checkbox" id="sidebar-checkbox"><!-- Toggleable sidebar --><div class="sidebar" id="sidebar">
        <div class="sidebar-item">
            <p>A reserved <a href="https://getnikola.com" target="_blank">Nikola</a> theme that places the utmost gravity on content with a hidden drawer. Made by <a href="https://twitter.com/mdo" target="_blank">@mdo</a> for Jekyll,
            ported to Nikola by <a href="https://twitter.com/ralsina" target="_blank">@ralsina</a>.</p>
        </div>
        
    <nav id="menu" role="navigation" class="sidebar-nav"><a class="sidebar-nav-item" href="archive.html">Archives</a>
        <a class="sidebar-nav-item" href="categories/index.html">Tags</a>
        <a class="sidebar-nav-item" href="rss.xml">RSS feed</a>
    
    
    </nav>
</div>

    <!-- Wrap is the content to shift when toggling the sidebar. We wrap the
         content to avoid any CSS collisions with our real content. -->
    <div class="wrap">
      <div class="masthead">
        <div class="container">
          
    <h3 id="brand" class="masthead-title">
      <a href="https://ilovesymposia.com/" title="I Love Symposia!" rel="home">I Love Symposia!</a>
    </h3>

        </div>
      </div>

      <div class="container content" id="content">
        

<div class="posts">
    <article class="post h-entry post-text"><header><h1 class="post-title p-name"><a href="posts/2015/05/04/the-scipy-ecosystem-belongs-to-everyone/" class="u-url">The SciPy ecosystem belongs to everyone</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn">Juan Nunez-Iglesias</span></p>
            <p class="dateline"><a href="posts/2015/05/04/the-scipy-ecosystem-belongs-to-everyone/" rel="bookmark"><time class="post-date published dt-published" datetime="2015-05-04T20:38:56+10:00" title="2015-05-04 20:38">2015-05-04 20:38</time></a></p>
                <p class="commentline">
        
    <a href="posts/2015/05/04/the-scipy-ecosystem-belongs-to-everyone/#disqus_thread" data-disqus-identifier="cache/posts/2015/05/04/the-scipy-ecosystem-belongs-to-everyone.html">Comments</a>


        </p>
</div>
    </header><div class="e-content entry-content">
    <div>
<p></p>
<p>I use Twitter favourites almost exclusively to mark posts that I know will be <em>useful</em> in some not-too-distant future; kind of like a Twitter Evernote. Recently I was looking through my list in search of <a href="http://www.pydanny.com/building-conda-packages-for-multiple-operating-systems.html">this excellent blog post</a> detailing how to build cross-platform binary distributions for <a href="http://conda.pydata.org/docs/">conda</a>.</p>
<p>I came across two other tweets from the EuroSciPy 2014 conference: <a href="https://twitter.com/ianozsvald/status/505739079590608896">this one</a> by Ian Ozsvald about his <a href="https://github.com/ianozsvald/ipython_memory_usage">IPython memory usage</a> profiler, right next to <a href="https://twitter.com/AlexChabotL/status/505746290412888065">this one</a> by Alexandre Chabot about Aaron O'Leary's <a href="https://github.com/aaren/notedown">notedown</a>. I'd forgotten that this was how I came across these two tools, but since then I have contributed code to both (<a href="https://github.com/ianozsvald/ipython_memory_usage/pull/1">1</a>, <a href="https://github.com/aaren/notedown/pull/15">2</a>). I'd met Ian at EuroSciPy 2013, but I've never met Aaron, yet nevertheless there is my code in the latest version of his notedown library.</p>
<p>How remarkable the open-source Python community has become. Talks from Python conferences are posted to YouTube, usually <em>as the conference is happening</em>. (Add to that plenty of live tweeting.) Thus, even when I can't attend the conferences, I can keep up with the latest open source libraries, from the other side of the world. And then I can grab the source code on GitHub, fiddle with it to my heart's content, and <a href="https://help.github.com/articles/using-pull-requests/">submit a pull request</a> to the author with my changes. After a short while, code that I wrote for my own utility is available for anyone else to use through PyPI or conda.</p>
<p>My point is: join us! <a href="http://www.astrobetter.com/blog/2014/03/10/the-whys-and-hows-of-licensing-scientific-code/">Make your code open source</a>, and conversely, when you need some functionality, don't reinvent the wheel. See if there's a library that <em>almost</em> meets your needs, and contribute!</p>
</div>
    </div>
    </article><article class="post h-entry post-text"><header><h1 class="post-title p-name"><a href="posts/2015/04/03/calling-out-scipy-on-diversity/" class="u-url">Calling out SciPy on diversity (even though it hurts)</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn">Juan Nunez-Iglesias</span></p>
            <p class="dateline"><a href="posts/2015/04/03/calling-out-scipy-on-diversity/" rel="bookmark"><time class="post-date published dt-published" datetime="2015-04-03T19:46:53+11:00" title="2015-04-03 19:46">2015-04-03 19:46</time></a></p>
                <p class="commentline">
        
    <a href="posts/2015/04/03/calling-out-scipy-on-diversity/#disqus_thread" data-disqus-identifier="cache/posts/2015/04/03/calling-out-scipy-on-diversity.html">Comments</a>


        </p>
</div>
    </header><div class="e-content entry-content">
    <div>
<p></p>
<p>Over the past few weeks, I've been heavily <a href="http://ilovesymposia.com/2015/03/23/go-to-scipy-2015/">promoting</a> the <a href="http://scipy2015.scipy.org/">SciPy conference</a>, a meeting about scientific programming in Python. I've been telling everyone who would listen that they should submit a talk abstract and go, because scientific programming is increasingly common in any scientist's work and SciPy massively improves how you do that.</p>
<p>I have also been guiltily ommitting that the speaker and attendee diversity at SciPy is shockingly bad. Last year, for example, 15% of attendees were women, and that was an improvement over the ratio three years ago, when just 3% (!!!) were women.</p>
<p>I rationalised continuing to promote this conference because there was <a href="https://www.youtube.com/watch?v=e0Z9-EuZJac">talk</a> from past organisers about making efforts to improve. (And indeed, the past three years have been on an upward trajectory.)<br></p>
<p>A couple of days ago, however, the full list of keynote speakers was announced, and lo and behold, it's three white guys. I have to acknowledge that they are extremely accomplished in the SciPy universe, and, if diversity were not more generally a problem at this conference and in tech in general, I wouldn't bat an eye. Excellent choice of speakers, really. Looking forward to it.</p>
<p>But diversity <i>is</i> a problem. It's an enormous problem. I'm inclined to call it catastrophic.</p>
<p>Let me try to quantify it. Men and women are equally capable scientific programmers. So out of a total pool of 100 potential SciPy attendees/contributors, 50 are women and 50 are men. Now, let's assume the male side of the community is working at near-optimum capacity, so, 50 of 50 those men are at SciPy. 15% of attendees being women means just 9 of the 50 potential female contributors are making it out to the conference (9/59 ≈ 15%). Or, slice it another way, a whopping (50 - 9) / 50 = 82% of women who could be contributing to SciPy are <i>missing</i>.</p>
<p>Now, think of where we would be if we took 82% of male science-Pythonistas and just erased their talks, their discussions, and their code contributions. The SciPy ecosystem would <i>suck</i>. Yet that is exactly how many coders are missing from our community.</p>
<p>Now, we can sit here and play <a href="http://www.feministe.us/blog/archives/2012/09/24/why-arent-there-more-women-at-stem-conferences-this-time-its-statistical/female-conference-speaker-bingo/">female conference speaker bingo</a> until the cows come home to roost, but that is missing the point: these are all only excuses for not doing enough. "Not my fault" is not good enough anymore. It is everyone's fault who does not make an active and prolonged effort to fix things.</p>
<p>The keynote speakers are an excellent place to make a difference, because you can eliminate all sorts of confounders. I have a certain sympathy, for example, for the argument that one should pick the best abstracts/scholarship recipients, rather than focus on race or gender. (Though the process should be truly blind to remove pervasive bias, as <a href="http://blogs.scientificamerican.com/unofficial-prognosis/2012/09/23/study-shows-gender-bias-in-science-is-real-heres-why-it-matters/">studies</a> and the experience of <a href="http://www.theguardian.com/women-in-leadership/2013/oct/14/blind-auditions-orchestras-gender-bias">orchestra auditions</a> have repeatedly shown.) For keynotes though, organisers are free to pursue any agenda they like. For example, they can make education a theme, and get Lorena Barba and Greg Wilson in, as they did last year.<br></p>
<p>Until the gender ratio begins to even remotely approach 1:1, diversity as an agenda should be a priority for the organisers. This doesn't mean invite the same number of women and men to give keynotes. This means <i>keep inviting qualified women until you have at least one confirmed female keynote speaker, </i>and preferably two. <i>Then, and only then,</i> you can look into inviting men.</p>
<p>Women have been <a href="http://www.timeshighereducation.co.uk/news/women-underrepresented-at-academic-conferences/2005015.article">found</a> to turn down conference invitations more often than men, irrespective of ability or accomplishment. I don't know why, but I suspect one reason is <a href="http://sciencecareers.sciencemag.org/career_magazine/previous_issues/articles/2010_02_12/science.opms.r1000084">lack of role models</a>, in the form of previous female speakers. That's why this keynote roster is so disappointing. There's tons of accomplished female Pythonistas out there, and there would be even more if we all made a concerted effort to improve things.</p>
<p>I don't want to retread the same territory that Jonathan Eisen (<a href="https://twitter.com/phylogenomics">@phylogenomics</a>) has already covered in "<a href="http://phylogenomics.blogspot.com.au/2015/03/calling-attention-to-meetings-with.html">Calling attention to meeting with skewed gender ratios, even when it hurts</a>". In particular, see that article for links to many others with ideas to improve gender ratios. But this is my contribution in the <i>exact same series: </i>I <i>love</i> SciPy. See my <a href="http://ilovesymposia.com/2014/07/17/scipy-2014-recap/">previous</a> <a href="http://ilovesymposia.com/2015/03/23/go-to-scipy-2015/">posts</a> for illustration.</p>
<p>Even looking back at my recent post, when I looked for a picture that I thought captured the collegial, collaborative feel of the conference, I unintentionally picked one featuring only men. This needs to improve, massively, if I'm going to keep supporting this conference. I really hope the organisers place diversity at the centre of their agenda for every decision going forward.</p>
<p><i>I thank <a href="https://twitter.com/phylogenomics">Jonathan Eisen</a>, <a href="https://twitter.com/aterrel">Andy Ray Terrel</a>, and <a href="https://twitter.com/wrightingapril">April Wright</a> for comments on earlier versions of this article.</i></p>
</div>
    </div>
    </article><article class="post h-entry post-text"><header><h1 class="post-title p-name"><a href="posts/2015/03/23/go-to-scipy-2015/" class="u-url">Go to SciPy 2015</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn">Juan Nunez-Iglesias</span></p>
            <p class="dateline"><a href="posts/2015/03/23/go-to-scipy-2015/" rel="bookmark"><time class="post-date published dt-published" datetime="2015-03-23T21:54:33+11:00" title="2015-03-23 21:54">2015-03-23 21:54</time></a></p>
                <p class="commentline">
        
    <a href="posts/2015/03/23/go-to-scipy-2015/#disqus_thread" data-disqus-identifier="cache/posts/2015/03/23/go-to-scipy-2015.html">Comments</a>


        </p>
</div>
    </header><div class="e-content entry-content">
    <div>
<p></p>
<a href="http://scipy2015.scipy.org">SciPy</a> is my favourite conference. My goal with this post is to convince someone to go who hasn't had that chance yet.
<p><a href="2015/03/reception.jpg"><img src="https://ilovesymposia.files.wordpress.com/2015/03/reception.jpg?w=450" alt="Photo by Ian Rees (from the SciPy 2012 conference website)" width="450" height="300" class="aligncenter size-medium wp-image-656"></a></p>
<h3>Why SciPy?</h3>

<p>Most scientists go to conferences in their own field: neuroscientists go to the monstrous Society for Neuroscience (SfN); Bioinformaticians go to RECOMB, ISMB, or PSB; and so on.</p>
<p>People go to these to keep up with the latest advances in their field, and often, to do a bit of networking.</p>
<p>SciPy is a different kind of conference. <em>It changes the way you do science.</em> You learn about the latest free and open source software to help you with your work. You learn to write functions and interfaces instead of scripts, and to write tests so you don't break your code. You also learn to contribute these to bigger projects, maximising the reach and impact of your work (see "sprints", below).</p>
<p>And you learn these things by <em>doing</em> them, with the people who are the best at this, rather than by reading books and blog posts. (Which maybe I shouldn't knock, since I'm writing a book about all this and you are reading my blog!)</p>
<p>Attendees to SciPy have scientific software in common, but come from diverse fields, including physics, pure maths, data visualisation, geosciences, climatology, and yes, biology and bioinformatics. Mingling with such a diverse group is a fantastic way to get your creative juices flowing!</p>
<p>The conference lasts a full week and is broken up into three parts: tutorials, main conference, and sprints.</p>
<h4>the tutorials</h4>

<p>With a few exceptions, you won't learn about your own field. But you will learn an enormous amount about <strong>tools</strong> that will help you be a better scientist. If you are a newbie to Python, you can go to the beginner tutorials track and learn about the fantastic scientific libraries available in Python. If you already use NumPy, SciPy, pandas, and the IPython notebook, you can go to the intermediate or advanced tracks, and learn new things about those. Even as an advanced SciPy user I still get tons of value from the tutorials. (Last year Min RK <a href="https://www.youtube.com/watch?v=U5mhpKkIx2Y">gave a <em>wild</em> demo</a> of IPython parallel's capabilities by crawling wikipedia remotely while building up a graph visualisation on his live notebook.) (Fast-forward to the 1h mark to see just the payoff.) Here's <a href="https://conference.scipy.org/scipy2014/schedule/tutorials/">last year's tutorial schedule</a> for an idea of what to expect.</p>
<h4>the main conference track</h4>

<p>You will also hear about the latest advances in the scientific libraries you know and use, and about libraries you didn't know about but will find useful (such as <a href="http://scikit-bio.org">scikit-bio</a>, <a href="http://yt-project.org">yt</a> or <a href="http://cmrivers.github.io/epipy/">epipy</a>). The main conference track features software advances written by presenters from many different fields. Hearing about these from the authors of the software lets you ask much different questions compared to hearing someone say, for example, "we used the Matlab image processing toolbox". If you ever had a feature request for your favourite library, or you wondered why they do something in a particular way, there's no better opportunity to get some closure.</p>
<p>The crosstalk between different fields is phenomenal. Hearing how a diverse set of people deal with their software problems really opens your mind to completely different approaches to what you had previously considered.</p>
<h4>the sprints</h4>

<p>Finally, there's two days of coding sprints. Even if you are a complete beginner in software development, <strong>do yourself a favour and participate in one of these</strong>.</p>
<p>Two and a half years after my first SciPy in 2012, I'm writing a scientific Python book for O'Reilly, and I can 100% trace it to participating in the scikit-image sprint that year. With their guidance, I wrote my <a href="https://github.com/scikit-image/scikit-image/pull/223">first ever GitHub pull request</a> and my first ever <a href="https://github.com/scikit-image/scikit-image/pull/223/files#diff-5d978a91c58ecc6a8275ccd24a6ddd3dR65">unit test</a>. Both were tiny and cute, and I would consider them trivial now, but that seed grew into massive improvements in my code-writing practice and many more contributions to open source projects.</p>
<p>And this is huge: now, instead of giving up when a software package doesn't do what I need it to do, I just look at the source code and figure out how I can add what I want. Someone else probably wants that functionality, and by putting it into a major software library instead of in my own code, I get it into the hands of many more users. It's a bit counterintuitive but there is nothing more gratifying than having some random person you've never met <a href="https://github.com/scikit-image/scikit-image/issues/679">complain when you break something</a>! This never happens when all your code is in your one little specialised repository containing functionality for your one paper.</p>
<h3>How SciPy</h3>

<p>The SciPy calls for tutorials, talks, posters, and its plotting contest are all out. There's specialised tracks and most of you reading this are probably interested in the <a href="http://microbe.net/2015/03/17/scipy-2015-computational-life-and-medical-sciences-mini-symposium-call-for-abstracts/">computational biology and medicine track</a>. It's taken me a while to write this post, so there's just one week left to submit something: the deadline is <strike>April 1st</strike> <strong>Update:</strong> the deadline for talks and posters has been extended to April 10th!</p>
<p>Even if you don't get something in, I encourage you to participate. Everything I said above still applies if you're not presenting. You might have a bit more trouble convincing your funders to pay for your travels, but if that's the case I encourage you to apply for <a href="http://scipy2015.scipy.org/ehome/115969/259279">financial assistance</a> from the conference.</p>
<p>I've <a href="http://ilovesymposia.com/2014/07/17/scipy-2014-recap/">written about</a> SciPy's diversity problem before, so I'm happy to report that this year there's specific <a href="http://scipy2015.scipy.org/ehome/115969/259279">scholarships</a> for women and minorities. (This may have been true last year, I forget.) <em>And</em>, awesomely, Matt Davis has <a href="http://penandpants.com/2015/03/12/scipy-2015-talk-help/">offered to help</a> first-time submitters with writing their proposals.</p>
<p>Give SciPy a try: <a href="http://www.scipy2015.scipy.org/eselectv2/frontend/index/115969">submit here</a> and/or <a href="https://www.etouches.com/ehome/115969/259272/">register here</a>. And feel free to <a href="mailto:juan.n@unimelb.edu.au">email me</a> or comment below if you have questions!</p>
<p><strong>Update:</strong> A colleague pointed out that I should also mention the awesomeness of the conference venue, so here goes: Austin in July is awesome. If you love the heat like I do, well, it doesn't get any better. If you don't, don't worry: the <a href="http://www.meetattexas.com">AT&amp;T Conference Center</a> AC is on friggin overdrive the whole time. Plus, there's some nearby <a href="http://en.wikipedia.org/wiki/Barton_Springs_Pool">cold springs</a> to swim in. The center itself is an excellent hotel and the conference organises massive discounts for attendees. There's a couple of great restaurants on-site; and the Mexican and Texas BBQ in the area are incredible — follow some Enthought and Continuum folks around to experience amazing food. Finally, Austin is a great city to bike in: last time I rented a road bike for the whole week from <a href="http://mellowjohnnys.com">Mellow Johnny's</a>, and enjoyed quite a few lunchtime and evening rides.</p>
</div>
    </div>
    </article><article class="post h-entry post-text"><header><h1 class="post-title p-name"><a href="posts/2015/03/08/experiences-porting-a-medium-sized-library-from-python-2-to-3/" class="u-url">Experiences porting a medium-sized library from Python 2 to 3</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn">Juan Nunez-Iglesias</span></p>
            <p class="dateline"><a href="posts/2015/03/08/experiences-porting-a-medium-sized-library-from-python-2-to-3/" rel="bookmark"><time class="post-date published dt-published" datetime="2015-03-08T23:33:34+11:00" title="2015-03-08 23:33">2015-03-08 23:33</time></a></p>
                <p class="commentline">
        
    <a href="posts/2015/03/08/experiences-porting-a-medium-sized-library-from-python-2-to-3/#disqus_thread" data-disqus-identifier="cache/posts/2015/03/08/experiences-porting-a-medium-sized-library-from-python-2-to-3.html">Comments</a>


        </p>
</div>
    </header><div class="e-content entry-content">
    <div>
<p></p>
<p>Prompted in part by some discussions with Ed Schofield, creator of <a href="http://python-future.org">python-future.org</a>, I've been going on a bit of a porting spree to Python 3. I just finished with my gala segmentation library. (Find it on <a href="https://github.com/janelia-flyem/gala">GitHub</a> and <a href="http://gala.readthedocs.org">ReadTheDocs</a>.) Overall, the process is nowhere near as onerous as you might think it is. Getting started really is the hardest part. If you have more than yourself as a user, you should definitely just get on with it and port.</p>
<p>The second hardest part is the testing. In particular, you will need to be careful with dictionary iteration, pickled objects, and file persistence in general. I'll go through these gotchas in more detail below.</p>
<p></p>
<h3>Reminder: the order of dictionary items is undefined</h3>
<p>This is one of those <em>duh</em> things that I forget over and over and over. In my porting, some tests that depended on a scikit-learn <code>RandomForest</code> object were failing. I assumed that there was some difference between the random seeding in Python 2 and Python 3, leading to slightly different models between the two versions of the random forest.</p>
<p>This was a <em>massive</em> red herring that took me forever to figure out. In actuality, the seeding was completely fine. However, gala uses networkx as its graph backend, which itself uses an adjacency dictionary to store edges. So when I asked for <code>graph.edges()</code> to get a set of training examples, I was getting the edges in a <em>random order</em> that was <em>deterministic</em> within Python 2.7: the edges returned were always in the <em>same</em> shuffled order. This went out the window when switching to Python 3.4, with the training examples now in a <em>different</em> order, resulting in a <em>different</em> random forest and thus a <em>different</em> learning outcome... And finally a failed test.</p>
<p>The solution <em>should</em> have been to use a classifier that is not sensitive to ordering of the training data. However, although many classifiers satisfy this property, in practice they suffer from slight numerical instability which is sufficient to throw the test results off between shufflings of the training data.</p>
<p>So I've trained a Naive Bayes classifier in Python 2.7, and which I then load up in Python 3.4 and check whether the parameters are <em>close</em> to a newly trained one. The actual classification results can differ slightly, and this becomes much worse in gala, where classification tasks are sequential, so a single misstep can throw off everything that comes after it.</p>
<h3>When pickling, remember to open files in binary mode</h3>

<p>I've always felt that the pickle module was deficient for not accepting filenames as input to <code>dump</code>. Instead, it takes an open, writeable file. This is all well and good but it turns out that you should always <a href="http://www.diveintopython3.net/serializing.html">open files in binary mode when using pickle</a>! I got this far without knowing that, surely an indictment of pickle's API!</p>
<p>Additionally, you'll have specify a <code>encoding='bytes'</code> when loading a Python 2 saved file in the Python 3 version of pickle.</p>
<h3>Even when you do, objects may not map cleanly between Python 2 and 3 (for some libraries)</h3>

<p>In Python 2:</p>
<pre class="code literal-block"><span></span><span class="k">[code lang=python]</span>
<span class="na">&amp;gt;&amp;gt;&amp;gt; from sklearn.ensemble import RandomForestClassifier as RF</span>
<span class="na">&amp;gt;&amp;gt;&amp;gt; rf</span> <span class="o">=</span> <span class="s">RF()</span>
<span class="na">&amp;gt;&amp;gt;&amp;gt; from sklearn.datasets import load_iris</span>
<span class="na">&amp;gt;&amp;gt;&amp;gt; iris</span> <span class="o">=</span> <span class="s">load_iris()</span>
<span class="na">&amp;gt;&amp;gt;&amp;gt; rf</span> <span class="o">=</span> <span class="s">rf.fit(iris.data, iris.target)</span>
<span class="na">&amp;gt;&amp;gt;&amp;gt; with open('rf', 'wb') as fout:</span>
<span class="na">...     pck.dump(r, fout, protocol</span><span class="o">=</span><span class="s">2)</span>
<span class="k">[/code]</span>
</pre>


<p>Then, in Python 3:</p>
<pre class="code literal-block"><span></span><span class="k">[code lang=python]</span>
<span class="na">&amp;gt;&amp;gt;&amp;gt; with open('rf', 'rb') as fin:</span>
<span class="na">...     rf</span> <span class="o">=</span> <span class="s">pck.load(fin, encoding='bytes')</span>
<span class="na">... </span>
<span class="na">---------------------------------------------------------------------------</span>
<span class="na">KeyError                                  Traceback (most recent call last)</span>
<span class="na">&lt;ipython-input-9-674ee92b354d&gt; in &lt;module&gt;()</span>
      <span class="na">1 with open('rf', 'rb') as fin:</span>
<span class="na">----&amp;gt; 2     rf</span> <span class="o">=</span> <span class="s">pck.load(fin, encoding='bytes')</span>
<span class="s">      3 </span>

<span class="na">/Users/nuneziglesiasj/anaconda/envs/py3k/lib/python3.4/site-packages/sklearn/tree/_tree.so in sklearn.tree._tree.Tree.__setstate__ (sklearn/tree/_tree.c:18115)()</span>

<span class="na">KeyError: 'node_count'</span>
<span class="k">[/code]</span>
</pre>


<h3>When all is said and done, your code will probably run slower on Python 3</h3>

<p>I have to admit: this just makes me angry. After <a href="https://github.com/janelia-flyem/gala/pull/39">a lot of hard work</a> ironing out all of the above kinks, gala's tests <a href="https://travis-ci.org/janelia-flyem/gala">run about 2x slower</a> in Python 3.4 than in 2.7. I'd heard quite a few times that Python 3 is slower than 2, but that's just ridiculous.</p>
<p>Nick Coghlan's <a href="http://python-notes.curiousefficiency.org/en/latest/python3/questions_and_answers.html">enormous Q&amp;A</a> has been <a href="http://sealedabstract.com/rants/python-3-is-fine/">cited</a> as required reading before complaining about Python 3. Well, I've read it (which took days), and I'm <em>still</em> angry that the CPython core development team are <a href="http://python-notes.curiousefficiency.org/en/latest/python3/questions_and_answers.html#why-doesn-t-this-limitation-really-bother-the-core-development-team">generally dismissive</a> of anyone wanting faster Python. Meanwhile, Google autocompletes "why is Python" with "so slow". And although Nick <a href="http://python-notes.curiousefficiency.org/en/latest/python3/questions_and_answers.html#what-about-insert-other-shiny-new-feature-here">asserts</a> that those of us complaining about this "misunderstand the perspective of conservative users", <a href="http://www.randalolson.com/2015/01/30/python-usage-survey-2014/">community surveys</a> show a whopping 40% of Python 2 users citing "no incentive" as the reason they don't switch.</p>
<h3>In conclusion...</h3>

<p>In the end, I'm glad I ported my code. I learned a few things, and I feel like a better Python "citizen" for having done it. But that's the point: those are pretty weak reasons. Most people just want to get their work done and move on. Why would they bother porting their code if it's not going to help them do that?</p>
</div>
    </div>
    </article><article class="post h-entry post-text"><header><h1 class="post-title p-name"><a href="posts/2015/02/23/clarifications-about-our-book-elegant-scipy-and-our-call-for-code-submissions/" class="u-url">Clarifications about our book, Elegant SciPy (and our call for code submissions)</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn">Juan Nunez-Iglesias</span></p>
            <p class="dateline"><a href="posts/2015/02/23/clarifications-about-our-book-elegant-scipy-and-our-call-for-code-submissions/" rel="bookmark"><time class="post-date published dt-published" datetime="2015-02-23T06:31:46+11:00" title="2015-02-23 06:31">2015-02-23 06:31</time></a></p>
                <p class="commentline">
        
    <a href="posts/2015/02/23/clarifications-about-our-book-elegant-scipy-and-our-call-for-code-submissions/#disqus_thread" data-disqus-identifier="cache/posts/2015/02/23/clarifications-about-our-book-elegant-scipy-and-our-call-for-code-submissions.html">Comments</a>


        </p>
</div>
    </header><div class="e-content entry-content">
    <div>
<p></p>
<h3>Short version</h3>
<p>Thank you to everyone who has already submitted, retweeted, and spread the word about <a href="http://ilovesymposia.com/2015/02/04/call-for-code-nominations-for-elegant-scipy/">our book, Elegant SciPy</a>! We are <strong>still looking for code submissions meeting these criteria</strong>:
- Submissions <strong>must</strong> use NumPy, SciPy, or a closely related library in a non-trivial way.
- Submissions <strong>must</strong> be (re)licensed as BSD, MIT, public domain, or something similarly liberal. (This is easy if you are the author.)
- Code should be satisfying in some way, such as speed, conciseness, broad applicability...
- Preferably, nominate <em>someone else's</em> code that impressed <em>you</em>.
- Include a scientific application on real data.</p>
<p>Submit by one of:
- Twitter: mention <a href="https://twitter.com/hdashnow">@hdashnow</a>, <a href="https://twitter.com/stefanvdwalt">@stefanvdwalt</a>, or <a href="https://twitter.com/jnuneziglesias">@jnuneziglesias</a>, or just use the hashtag #ElegantSciPy;
- Email: <a href="mailto:stefanv(at)berkeley.edu">Stéfan van der Walt</a>, <a href="mailto:juan.n@unimelb.edu.au">Juan Nunez-Iglesias</a>, or <a href="mailto:harriet.dashnow@unimelb.edu.au">Harriet Dashnow</a>; or
- GitHub: create a new issue <a href="https://github.com/HarrietInc/elegant-scipy-submissions/issues">here</a>.
(All submissions will be credited as "written by" and "nominated by".)</p>
<h3>Long version</h3>

<p>A big thank you to everyone that has submitted code, retweeted, posted on mailing lists, and otherwise spread the word about the book! We're still pretty far from a book-length title though. I was also a bit vague about the kinds of submissions we wanted. I'll elaborate a bit on each of the above points:</p>
<h4>NumPy and SciPy use</h4>

<p>Some excellent submissions did not use the SciPy library, but rather did amazing things with the Python standard library. I should have mentioned that the book will be specifically focused on the NumPy and SciPy libraries. That's just the scope of the book that O'Reilly has contracted us to write. Therefore, although we might try to fit examples of great Python uses into a chapter, they are not suitable to be the centerpieces.</p>
<p>We will make some exceptions, for example for very closely related libraries such as pandas and scikit-learn. But, generally, the scope is SciPy the library.</p>
<h4>Licensing</h4>

<p>This one's pretty obvious. We can't use your submission if it's under a restrictive license. And we don't want to publish GPL-licensed code, which could force our readers to GPL-license their own code when using it. For more on this, see <a href="http://choosealicense.com">Choose A License</a>, as well as Jake Vanderplas's excellent <a href="http://www.astrobetter.com/the-whys-and-hows-of-licensing-scientific-code/">blog post</a> encouraging the use of the BSD license for scientific code.</p>
<p>Note that the author of a piece of code is free to relicense as they choose, so if you think some GPL code is perfect and might be amenable to relicensing, do let us know about it!</p>
<h4>Submitting someone else's code</h4>

<p>I suspect that a lot of people are shy about submitting their own code. Two things should alleviate this. First, you can now submit via email, so you don't have to be public about the self-promotion. (Not that there's anything wrong with that, but I know I sometimes struggle with it.) And second, I want to explicitly state that we <em>prefer it</em> if you submit <em>others'</em> code. This is not to discourage self-promotion, but to drive code quality even higher. It's a high bar to convince ourselves that our code is worthy of being called elegant, but it takes another level entirely to find someone else's code elegant! (The usual reaction when reading other people's code is more like, "and what the <em>#$&amp;^</em> is going on <em>here????</em>") So, try to think about times you saw great code during a code review, reading a blog post, or while grokking and fiddling with someone else's library.</p>
<h4>Data</h4>

<p>Beautiful code is kind of a goal unto itself, but we really want to demonstrate how <em>useful</em> SciPy is in real scientific analysis. Therefore, although cute examples on synthetic data can illustrate quite well what a piece of code does, it would be extremely useful for us if you can point to examples with real scientific data behind them.</p>
<p>Thank you, and we hope to hear from you!</p>
</div>
    </div>
    </article><article class="post h-entry post-text"><header><h1 class="post-title p-name"><a href="posts/2015/02/04/call-for-code-nominations-for-elegant-scipy/" class="u-url">Call for code nominations for Elegant SciPy!</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn">Juan Nunez-Iglesias</span></p>
            <p class="dateline"><a href="posts/2015/02/04/call-for-code-nominations-for-elegant-scipy/" rel="bookmark"><time class="post-date published dt-published" datetime="2015-02-04T16:23:31+11:00" title="2015-02-04 16:23">2015-02-04 16:23</time></a></p>
                <p class="commentline">
        
    <a href="posts/2015/02/04/call-for-code-nominations-for-elegant-scipy/#disqus_thread" data-disqus-identifier="cache/posts/2015/02/04/call-for-code-nominations-for-elegant-scipy.html">Comments</a>


        </p>
</div>
    </header><div class="e-content entry-content">
    <div>
<p></p>
<strong>Update:</strong> See the also the <a href="http://ilovesymposia.com/2015/02/23/clarifications-about-our-book-elegant-scipy-and-our-call-for-code-submissions/">clarifications</a> to this post, and submit code by creating an issue in our <a href="https://github.com/HarrietInc/elegant-scipy-submissions">GitHub repo</a>!
<p>It's official! Harriet Dashnow, Stéfan van der Walt, and I will be writing an O'Reilly book about the SciPy library and the surrounding ecosystem. The book is called Elegant SciPy, and is intended to teach SciPy to fledgling Pythonistas, guided by the most elegant SciPy code examples we can find.</p>
<p>So, if you recently came across scientific Python code that made you go "Wow!" with its elegance, simplicity, cleverness, or power, please point us to it! As an example, have a look at Vighnesh Birodkar's code to build a region adjacency graph from an n-dimensional image, which I highlighted previously <a href="http://ilovesymposia.com/2014/06/24/a-clever-use-of-scipys-ndimage-generic_filter-for-n-dimensional-image-processing/">here</a>.</p>
<p>Each chapter will have one or two snippets that we will work towards. Each of these will be credited as "written by/nominated by", and needs to be published under a permissive license such as MIT, BSD, or public domain to be considered for inclusion. We would especially like nominations in the following categories:</p>
<ul>
<li>statistics (using scipy.stats)</li>
<li>image processing and computer vision</li>
<li>Fourier transforms</li>
<li>sparse matrix operations</li>
<li>eigendecomposition and linear algebra</li>
<li>optimization</li>
<li>streaming data analysis</li>
<li>spatial and geophysical data analysis</li>
</ul>
<p>We'll also consider other parts of the SciPy library and ecosystem.</p>
<p><strong>We invite you to submit code snippets for inclusion in the book.</strong> We'd also appreciate a small description of about one paragraph explaining what the code is used for and why you think it's elegant, even though this is often self-evident. =)</p>
<h4>How to submit</h4>

<ul>
<li>Reach out to any of us on Twitter: <a href="https://twitter.com/hdashnow">@hdashnow</a>, <a href="https://twitter.com/stefanvdwalt">@stefanvdwalt</a>, <a href="https://twitter.com/jnuneziglesias">@jnuneziglesias</a>.</li>
<li>Or just comment on the blog post, below.</li>
<li>
<strong>Update</strong>: Or, the preferred method, add an issue in our <a href="https://github.com/HarrietInc/elegant-scipy-submissions">GitHub repository</a>!</li>
</ul>
<p>Thank you,</p>
<p>Juan, Harriet, and Stéfan.</p>
</div>
    </div>
    </article><article class="post h-entry post-text"><header><h1 class="post-title p-name"><a href="posts/2015/01/04/some-things-i-learned-while-building-a-site-on-github-pages/" class="u-url">Some things I learned while building a site on GitHub Pages</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn">Juan Nunez-Iglesias</span></p>
            <p class="dateline"><a href="posts/2015/01/04/some-things-i-learned-while-building-a-site-on-github-pages/" rel="bookmark"><time class="post-date published dt-published" datetime="2015-01-04T05:09:10+11:00" title="2015-01-04 05:09">2015-01-04 05:09</time></a></p>
                <p class="commentline">
        
    <a href="posts/2015/01/04/some-things-i-learned-while-building-a-site-on-github-pages/#disqus_thread" data-disqus-identifier="cache/posts/2015/01/04/some-things-i-learned-while-building-a-site-on-github-pages.html">Comments</a>


        </p>
</div>
    </header><div class="e-content entry-content">
    <div>
<p></p>
<p>Understatement: I'm not much of a web developer. However, we all have to become a little bit versed in web-dev if we want to publish things these days. GitHub Pages makes it really easy to publish a site (check out the <a href="https://pages.github.com">official guide</a>, and Thinkful's truly excellent <a href="http://www.thinkful.com/learn/a-guide-to-using-github-pages/">interactive getting started guide</a>).</p>
<p>If you just want to publish a static set of html files using absolute paths, you'll be fine. However, Pages uses <a href="http://jekyllrb.com">Jekyll</a>, a server that can transform collections of Markdown, HTML, and other files into full-fledged websites. The process is definitely full of gotchas, though, and you'll run into issues for anything other than single pages. I'm making this list for my own future reference, and so that I can finally close the umpteen tabs I have open on the topic! But I hope someone else will find it useful.</p>
<p></p>
<h3>1. Jekyll lets you write Markdown, but it's not GitHub-Flavored Markdown</h3>
<p><strong>Update:</strong> This is no longer true. You can use GFM (or something very similar to it) using <code>kramdown</code> as the renderer.</p>
<p>You might assume, since you are writing markdown hosted on GitHub, that <a href="https://github.github.com/github-flavored-markdown/">GitHub-Flavored Markdown (GFM)</a> would work. But you would be wrong. You can configure Jekyll on GH-Pages to use either <a href="https://github.com/vmg/redcarpet">Redcarpet</a> or <a href="http://kramdown.gettalong.org/syntax.html">Kramdown</a>, both of which have significant syntactic differences with GFM. For example, fenced code blocks <a href="http://kramdown.gettalong.org/syntax.html#fenced-code-blocks">use <code>~~~</code> in Kramdown</a>, instead of GFM's <code>```</code>.</p>
<p>Anyway, to set the Markdown engine used to render your site, set the <code>markdown</code> attribute in the Jekyll configuration file, <code>_config.yml</code>, placed at the root of your repo:</p>
<pre><code>
# Dependencies
markdown:    kramdown
highlighter: pygments
</code></pre>

<h3>2. ... And syntax highlighting doesn't work properly in Jekyll on GH-Pages</h3>

<p><strong>Update:</strong> This has also been fixed.</p>
<p>Even using the correct delimiters for Kramdown is not good enough to get your code snippets to render properly, because of <a href="https://github.com/jekyll/jekyll/issues/2709">this</a> and <a href="https://github.com/jekyll/jekyll/issues/2715">this</a>. So, instead of using backtick- or tilde-fenced blocks, you have to use slightly clunkier Jekyll "liquid tags", which have the following fairly ugly syntax:</p>
<pre>
{% highlight python %}
    from scipy import ndimage as nd
{% endhighlight %}
</pre>

<h3>3. Linking is a bit weird</h3>

<p>My expectation when I started working with this was that I could write in markdown, and link directly to other markdown files using relative URLs, and Jekyll would just translate these to HTML links when serving the site. This is how <a href="http://www.mkdocs.org">MkDocs</a> and local viewers such as <a href="http://marked2app.com">Marked 2</a> work, for example.</p>
<p>That's not at all how Jekyll does things. <strong>Every markdown page that you want served needs to have a header called the <a href="http://jekyllrb.com/docs/frontmatter/">Front Matter</a>.</strong> Other markdown files will be publicly accessible but they will be treated as plaintext files.</p>
<p><em>How</em> the markdown files get served also varies depending on your <code>_config.yml</code> and the <code>permalink</code> variable. The annoying thing is that the <a href="https://github.com/jekyll/jekyll/issues/1293">documentation</a> on this is entirely centred on blog posts, with little indication about what happens to top-level pages. In the case of the <code>pretty</code> setting, <code>filename.markdown</code>, for example, gets served at <code>base.url/filename/</code>, <em>unless</em> you set the page's <code>permalink</code> attribute manually in the Front Matter, for example to <code>title.html</code> or <code>title/</code>.</p>
<h3>4. Relative paths don't work on GH-Pages</h3>

<p>This is a biggy, and super-annoying, because even if you follow GitHub's <a href="https://help.github.com/articles/using-jekyll-with-pages/">guide for local development</a>, <em>all</em> your style and layout files will be missing once you deploy to Pages.</p>
<p>The problem is that Jekyll assumes that your site lives at the root URL (i.e. <code>username.github.io</code>), when it is in fact in <code>username.github.io/my-site</code>.</p>
<p>The solution was developed by Matt Swensen <a href="https://github.com/jekyll/jekyll/issues/332#issuecomment-18952908">here</a>: You need to set a <code>baseurl</code> variable in your <code>_config.yml</code>. This should be your repository name. For example, if you post your site to the <code>gh-pages</code> branch of the repository <code>https://github.com/username/my-jekyll-site</code>, which gets served at <code>username.github.io/my-jekyll-site/</code>, you should set <code>baseurl</code> to <code>my-jekyll-site</code>.</p>
<p>Once you've set the <code>baseurl</code>, you <em>must</em> use it when linking to CSS and posts, using liquid tags: <code>{{ site.baseurl }}/path/to/style.css</code> and <code>{{ site.baseurl }}{{ post.url }}</code>.</p>
<p>When you're testing the site locally, you can either manually configure the <code>baseurl</code> tag to the empty string (<code>bundle exec jekyll serve --baseurl ''</code>), so you can navigate to 0.0.0.0:4000 as you would normally with Jekyll; or you can <a href="http://blog.parkermoore.de/2014/04/27/clearing-up-confusion-around-baseurl">leave it untouched</a> and navigate to 0.0.0.0:4000/my-jekyll-site, mirroring the structure of GitHub Pages.</p>
<p><strong>This is all documented in a specific Jekyll docs page</strong> that is unfortunately buried after <em>a lot</em> of other, less-relevant Jekyll docs. As a reference, look <a href="http://jekyllrb.com/docs/github-pages/">here</a>, and search for "Project Page URL Structure".</p>
<h3>5. You <em>must</em> provide an index.html or index.md file</h3>

<p>Otherwise you get a rather unhelpful <a href="https://github.com/jekyll/jekyll/issues/1293">Forbidden Error</a>.</p>
<h3>And 6. There is a "console" syntax highlighter for bash sessions</h3>

<p>I'd always used "bash" as my highlighter whenever I wanted to show a shell session, but it turns out that "console" does a slightly better job, by highlighting <code>$</code> prompts and commands, while graying output slightly. Christian Jann, who has written his own more-specific highlighter, ShellSession, has a <a href="http://www.jann.cc/pygments-shell-session-lexer-demo/shell_code_comparison.html">nice comparison</a> of the three. Although ShellSession is cool, I couldn't get it to work out-of-the-box on Jekyll.</p>
<hr>
<p>That should be enough to get you started... (Don't forget to check out <a href="http://www.thinkful.com/learn/a-guide-to-using-github-pages/">Thinkful's guide</a>!) Happy serving!</p>
</div>
    </div>
    </article><article class="post h-entry post-text"><header><h1 class="post-title p-name"><a href="posts/2014/10/27/continuous-integration-in-python-7-some-helper-tools-and-final-thoughts/" class="u-url">Continuous integration in Python, 7: some helper tools and final thoughts</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn">Juan Nunez-Iglesias</span></p>
            <p class="dateline"><a href="posts/2014/10/27/continuous-integration-in-python-7-some-helper-tools-and-final-thoughts/" rel="bookmark"><time class="post-date published dt-published" datetime="2014-10-27T07:03:52+11:00" title="2014-10-27 07:03">2014-10-27 07:03</time></a></p>
                <p class="commentline">
        
    <a href="posts/2014/10/27/continuous-integration-in-python-7-some-helper-tools-and-final-thoughts/#disqus_thread" data-disqus-identifier="cache/posts/2014/10/27/continuous-integration-in-python-7-some-helper-tools-and-final-thoughts.html">Comments</a>


        </p>
</div>
    </header><div class="e-content entry-content">
    <div>
<p>Almost everything I know about good Python development I've learned from <a href="https://github.com/stefanv">Stéfan van der Walt</a>, <a href="https://github.com/tonysyu">Tony Yu</a>, and the rest of the <a href="https://github.com/scikit-image/scikit-image/graphs/contributors">scikit-image team</a>. But a few weeks ago, I was trying to emulate the scikit-image CI process for my own project: <a href="https://github.com/jni/cellom2tif">cellom2tif</a>, a tool to liberate images from a <a href="http://www.openmicroscopy.org/site/support/bio-formats5/formats/cellomics.html">rather useless</a> proprietary format. (I consider this parenthetical comment sufficient fanfare to announce the <a href="https://pypi.python.org/pypi/cellom2tif/0.2.0">0.2 release!</a>) As I started copying and editing config files, I found that even from a complete template, getting started was not very straightforward. First, scikit-image has much more complicated requirements, so that a lot of the <code>.travis.yml</code> file was just noise for my purposes. And second, as detailed in the previous posts, a lot of the steps are not found or recorded anywhere in the repository, but rather must be navigated to on the webpages of GitHub, Travis, and Coveralls. I therefore decided to write this series as both a notetaking exercise and a guide for future CI novices. (Such as future me.)
.. has_math: no
.. status: published
.. wp-status: publish
--&gt;</p>
<p></p>
<p>It's time to draw my "continuous integration in Python" series to a close. This final post ties all six previous posts together and is the preferred write-up to share more widely and on which to provide feedback.</p>
<p>Almost everything I know about good Python development I've learned from <a href="https://github.com/stefanv">Stéfan van der Walt</a>, <a href="https://github.com/tonysyu">Tony Yu</a>, and the rest of the <a href="https://github.com/scikit-image/scikit-image/graphs/contributors">scikit-image team</a>. But a few weeks ago, I was trying to emulate the scikit-image CI process for my own project: <a href="https://github.com/jni/cellom2tif">cellom2tif</a>, a tool to liberate images from a <a href="http://www.openmicroscopy.org/site/support/bio-formats5/formats/cellomics.html">rather useless</a> proprietary format. (I consider this parenthetical comment sufficient fanfare to announce the <a href="https://pypi.python.org/pypi/cellom2tif/0.2.0">0.2 release!</a>) As I started copying and editing config files, I found that even from a complete template, getting started was not very straightforward. First, scikit-image has much more complicated requirements, so that a lot of the <code>.travis.yml</code> file was just noise for my purposes. And second, as detailed in the previous posts, a lot of the steps are not found or recorded anywhere in the repository, but rather must be navigated to on the webpages of GitHub, Travis, and Coveralls. I therefore decided to write this series as both a notetaking exercise and a guide for future CI novices. (Such as future me.)</p>
<p>To recap, here are my six steps to doing continuous integration in Python with pytest, Travis, and Coveralls:</p>
<p></p>
<ul>
<li><a href="http://ilovesymposia.com/2014/10/01/continuous-integration-0-automated-tests-with-pytest/">Write tests</a></li>
<li><a href="http://ilovesymposia.com/2014/10/02/continuous-integration-1-test-coverage/">Measure test coverage</a></li>
<li><a href="http://ilovesymposia.com/2014/10/13/continuous-integration-in-python-3-set-up-your-test-configuration-files/">Set up config files</a></li>
<li><a href="http://ilovesymposia.com/2014/10/15/continuous-integration-in-python-4-set-up-travis-ci/">Turn on Travis-CI</a></li>
<li><a href="http://ilovesymposia.com/2014/10/15/continuous-integration-in-python-5-report-test-coverage-using-coveralls/">Turn on Coveralls</a></li>
<li><a href="http://ilovesymposia.com/2014/10/17/continuous-integration-in-python-6-show-off-your-work/">Badge your repo</a></li>
</ul>
<p>If you do all of the above at the beginning of your projects, you'll be in a really good place one, two, five years down the line, when many academic projects grow far beyond their original scope in unpredictable ways and end up with much broken code. (See this <a href="http://www.nature.com/news/2010/101013/full/467775a.html">wonderful editorial</a> by Zeeya Merali for much more on this topic.)</p>
<h3>Reducing the boilerplate with PyScaffold</h3>

<p>But it's a lot of stuff to do for every little project. I was about to make myself some minimal <code>setup.cfg</code> and <code>.travis.yml</code> template files so that I could have these ready for all new projects, when I remembered <a href="http://pyscaffold.readthedocs.org/en/latest/index.html">PyScaffold</a>, which sets up a Python project's basic structure automatically (<code>setup.py</code>, <code>package_name/<strong>init</strong>.py</code>, etc.). Sure enough, PyScaffold has a <a href="http://pyscaffold.readthedocs.org/en/latest/features.html#unittest-coverage"><code>--with-travis</code> option</a> that implements <em>all</em> my recommendations, including pytest, Travis, and Coveralls. If you set up your projects with PyScaffold, you'll just have to turn on Travis-CI on your GitHub repo admin and Coveralls on coveralls.io, and you'll be good to go.</p>
<h3>When Travises attack</h3>

<p>I've made a fuss about how wonderful Travis-CI is, but it breaks more often than I'd like. You'll make some changes locally, and ensure that the tests pass, but when you push them to GitHub, Travis fails. This can happen for various reasons:</p>
<ul>
<li>your environment is different (e.g. NumPy versions differ between your local build and Travis's VMs).</li>
<li>you're testing a function that depends on random number generation and have failed to set the seed.</li>
<li>you depend on some web resource that was temporarily unavailable when you pushed.</li>
<li>Travis has updated its VMs in some incompatible way.</li>
<li>you have more memory/CPUs locally than Travis allows.</li>
<li>some other, not-yet-understood-by-me reason.</li>
</ul>
<p>Of these, the first three are acceptable. You can use <a href="http://conda.pydata.org/docs/">conda</a> to match your environments both locally and on Travis, and you should <a href="http://blog.amvtek.com/posts/2014/Jun/20/making-good-use-of-random-in-your-python-unit-tests/">always set the seed</a> for randomised tests. For network errors, Travis provides a special function, <a href="http://blog.travis-ci.com/2013-05-20-network-timeouts-build-retries/"><code>travis_retry</code></a>, that you can prefix your commands with.</p>
<p>Travis VM updates should theoretically be benign and not cause any problems, but, in recent months, they have been a <a href="https://github.com/scikit-image/scikit-image/pull/1090#issuecomment-50983376">significant</a> <a href="https://github.com/scikit-image/scikit-image/pull/1100#issuecomment-53651935"> source</a> <a href="https://github.com/scikit-image/scikit-image/pull/1189#issuecomment-58500996">of</a> <a href="https://github.com/scikit-image/scikit-image/pull/1189#issuecomment-58610095">pain</a> for the scikit-image team: every monthly update by Travis broke our builds. That's disappointing, to say the least. For simple builds, you really shouldn't run into this. But for major projects, this is an unnecessary source of instability.</p>
<p>Further, Travis VMs don't have unlimited memory and disk space for your builds (naturally), but <a href="https://github.com/scikit-image/scikit-image/pull/1199#issuecomment-59448250">the limits are not strictly defined</a> (unnaturally). This means that builds requiring "some" memory or disk space randomly fail. Again, disappointing. Travis could, for example, guarantee some minimal specs that everyone could program against — and request additional space either as special exemptions or at a cost.</p>
<p>Finally, there's the weird failures. I don't have any examples on hand but I'll just note that sometimes Travis builds fail, where your local copy works fine every single time. Sometimes <a href="http://stackoverflow.com/a/21727193">rebuilding</a> fixes things, and other times you have to change some subtle but apparently inconsequential thing before the build is fixed. These would be mitigated if Travis allowed you to clone their VM images so you could run them on a local VM or on your own EC2 allocation.</p>
<p><a href="2014/10/screen-shot-2014-10-27-at-7-57-36-pm.png"><img src="https://ilovesymposia.files.wordpress.com/2014/10/screen-shot-2014-10-27-at-7-57-36-pm.png?w=450" alt="Heisenbug" width="450" height="84" class="size-large wp-image-597"></a> A too-common Travis occurrence: randomly failing tests</p>
<p>In all though, Travis is a fantastic resource, and you shouldn't let my caveats stop you from using it. They are just something to keep in mind before you pull <em>all</em> your hair out.</p>
<h3>The missing test: performance benchmarks</h3>

<p>Testing helps you maintain the correctness of your code. However, as Michael Droettboom <a href="https://www.youtube.com/watch?v=OsxJ5O6h8s0">eloquently argued</a> at SciPy 2014, all projects are prone to feature creep, which can progressively slow code down. <a href="https://spacetelescope.github.io/asv/">Airspeed Velocity</a> is to benchmarks what pytest is to unit tests, and allows you to monitor your project's speed over time. Unfortunately, benchmarks are a different beast to tests, because you need to keep the testing computer's specs and load constant for each benchmark run. Therefore, a VM-based CI service such as Travis is out of the question.</p>
<p>If your project has <em>any</em> performance component, it may well be worth investing in a dedicated machine only to run benchmarks. The machine could monitor your GitHub repo for changes and PRs, check them out when they come in, run the benchmarks, and report back. I have yet to do this for any of my projects, but will certainly consider this strongly in the future.</p>
<h3>Some reservations about GitHub</h3>

<p>The above tools all work great as part of GitHub's <a href="https://help.github.com/articles/using-pull-requests/">pull request (PR)</a> development model. It's a model that is easy to grok, works well with new programmers, and has driven massive growth in the open-source community. Lately, I recommend it with a bit more trepidation than I used to, because it does have a few high-profile detractors, notably Linux and git creator <a href="https://github.com/torvalds/linux/pull/17#issuecomment-5654674">Linus Torvalds</a>, and OpenStack developer <a href="https://julien.danjou.info/blog/2013/rant-about-github-pull-request-workflow-implementation">Julien Danjou</a>. To paraphrase Julien, there are two core problems with GitHub's chosen workflow, both of which are longstanding and neither of which shows any sign of improving.</p>
<p>First, <strong>comments on code diffs are buried by subsequent changes</strong>, whether the changes are a rebase or they simply change the diff. This makes it very difficult for an outside reviewer to assess what discussion, if any, resulted in the final/latest design of a PR. This could be a fairly trivial fix (colour-code outdated diffs, rather than hiding them), so I would love to see some comments from GitHub as to what is taking so long.</p>
<p><a href="2014/10/screen-shot-2014-10-27-at-5-06-29-pm.png"><img src="https://ilovesymposia.files.wordpress.com/2014/10/screen-shot-2014-10-27-at-5-06-29-pm.png?w=660" alt="GitHub's hidden PR comments" width="660" height="127" class="size-large wp-image-596"></a> Expect to see a lot of these when using pull requests.</p>
<p>Second, <strong>bisectability is broken by fixup commits.</strong> The GitHub development model is not only geared towards small, incremental commits being piled on to a history, but it actively encourages these with their per-commit badging of a user's <a href="https://help.github.com/articles/viewing-contributions-on-your-profile-page/#contributions-calendar">contribution calendar</a>. Fixup commits make bug hunting with <a href="http://git-scm.com/docs/git-bisect">git bisect</a> more difficult, because some commits will not be able to run a test suite at all. This could be alleviated by considering only commits merging GitHub PRs, whose commit message start with <code>Merge pull request #</code>, but I don't know how to get git to do this automatically (ideas welcome in the comments).</p>
<p>I disagree with Julien that there is "no value in the social hype [GitHub] brings." In fact, GitHub has dramatically improved my coding skills, and no doubt countless others'. For many, it is their first experience with code review. Give credit where it is due: GitHub is driving the current, enormous wave of open-source development. But there is no doubt it needs improvement, and it's sad to see GitHub's developers apparently ignoring their critics. I hope the latter will be loud enough soon that GitHub will have no choice but to take notice.</p>
<h3>Final comments</h3>

<p>This series, including this post, sums up my current thinking on CI in Python. It's surely incomplete: I recently came across a curious "Health: 88%" badge on Mitchell Stanton-Cook's <a href="https://github.com/mscook/BanzaiDB">BanzaiDB</a> README. Clicking it took me to the project's <a href="https://landscape.io/github/mscook/BanzaiDB/52">landscape.io page</a>, which appears to do for coding style what Travis does for builds/tests and Coveralls does for coverage. How it measures "style" is not yet clear to me, but it might be another good CI tool to keep track of. Nevertheless, since it's taken me a few years to get to this stage in my software development practice, I hope this series will help other scientists get there faster.</p>
<p>If any more experienced readers think any of my advice is rubbish, please speak up in the comments! I'll update the post(s) accordingly. CI is a big rabbit hole and I'm still finding my way around.</p>
</div>
    </div>
    </article><article class="post h-entry post-text"><header><h1 class="post-title p-name"><a href="posts/2014/10/17/continuous-integration-in-python-6-show-off-your-work/" class="u-url">Continuous integration in Python, 6: Show off your work</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn">Juan Nunez-Iglesias</span></p>
            <p class="dateline"><a href="posts/2014/10/17/continuous-integration-in-python-6-show-off-your-work/" rel="bookmark"><time class="post-date published dt-published" datetime="2014-10-17T06:37:37+11:00" title="2014-10-17 06:37">2014-10-17 06:37</time></a></p>
                <p class="commentline">
        
    <a href="posts/2014/10/17/continuous-integration-in-python-6-show-off-your-work/#disqus_thread" data-disqus-identifier="cache/posts/2014/10/17/continuous-integration-in-python-6-show-off-your-work.html">Comments</a>


        </p>
</div>
    </header><div class="e-content entry-content">
    <div>
<p></p>
<p>We're finally ready to wrap up this topic. By now you can:</p>
<p></p>
<ul>
<li><a href="http://ilovesymposia.com/2014/10/01/continuous-integration-0-automated-tests-with-pytest/">Automatically test your code with pytest</a></li>
<li><a href="http://ilovesymposia.com/2014/10/02/continuous-integration-1-test-coverage/">Measure your test coverage</a></li>
<li><a href="http://ilovesymposia.com/2014/10/13/continuous-integration-in-python-3-set-up-your-test-configuration-files/">Configure your tests</a></li>
<li>Use Travis-CI to <a href="http://ilovesymposia.com/2014/10/15/continuous-integration-in-python-4-set-up-travis-ci/">run your tests automatically with each push</a>
</li>
<li>Use Coveralls to <a href="http://ilovesymposia.com/2014/10/15/continuous-integration-in-python-5-report-test-coverage-using-coveralls/">measure your coverage with each push</a>
</li>
</ul>
<p>But, much as exercise is wasted if your bathroom scale doesn't automatically tweet about it, all this effort is for naught if visitors to your GitHub page can't see it!</p>
<p>Most high-profile open-source projects these days advertise their CI efforts. Above, I cheekily called this showing off, but it's truly important: anyone who lands on your GitHub page is a potential user or contributor, and if they see evidence that your codebase is stable and well-tested, they are more likely to stick around.</p>
<p>Badging your README is easy. (You <em>do</em> have a README, <a href="posts/2014/10/17/continuous-integration-in-python-6-show-off-your-work/">don't you?</a>) In Travis, go to your latest build. Near the top right, click on the "build passing" badge:</p>
<p><img src="2014/10/screen-shot-2014-10-17-at-8-20-00-pm.png" alt="Travis-CI badge"></p>
<p>You'll get an overlay, with a pull-down menu for all the different options for getting the badge. You can grab the image URL, or you can directly grab the Markdown to put into your markdown-formatted README, or a bunch of other options, including RST:</p>
<p><img src="2014/10/screen-shot-2014-10-17-at-8-27-05-pm.png" alt="Travis-CI badge URLs"></p>
<p>Just copy and paste the appropriate code and add it to your README file wherever you please.</p>
<p>Meanwhile, on your repository's Coveralls page, on the right-hand side, you will find another pull-down menu with the appropriate URLs:</p>
<p><img src="2014/10/screen-shot-2014-10-17-at-8-28-24-pm.png" alt="Coveralls badge URLs"></p>
<p>Again, just grab whichever URL is appropriate for your needs (I prefer Markdown-formatted READMEs), and add it to your README, usually next to the Travis badge.</p>
<p>And you're done! You've got <a href="http://ilovesymposia.com/2014/10/01/continuous-integration-0-automated-tests-with-pytest/">automated tests</a>, <a href="http://ilovesymposia.com/2014/10/02/continuous-integration-1-test-coverage/">tons of test coverage</a>, you're running everything correctly thanks to <a href="http://ilovesymposia.com/2014/10/13/continuous-integration-in-python-3-set-up-your-test-configuration-files/">configuration files</a>, and all this is getting run on demand thanks to <a href="http://ilovesymposia.com/2014/10/15/continuous-integration-in-python-4-set-up-travis-ci/">Travis</a> and <a href="http://ilovesymposia.com/2014/10/15/continuous-integration-in-python-5-report-test-coverage-using-coveralls/">Coveralls</a>. And thanks to badging, the whole world knows about it:</p>
<p><img src="2014/10/screen-shot-2014-10-18-at-12-33-37-am.png" alt="Readme badges"></p>
<p><strong>Update:</strong> <a href="https://ilovesymposia.com/2014/10/27/continuous-integration-in-python-7-some-helper-tools-and-final-thoughts/">Volume 7: Some final thoughts, tips, and tools.</a></p>
</div>
    </div>
    </article><article class="post h-entry post-text"><header><h1 class="post-title p-name"><a href="posts/2014/10/15/continuous-integration-in-python-5-report-test-coverage-using-coveralls/" class="u-url">Continuous integration in Python, 5: report test coverage using Coveralls</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn">Juan Nunez-Iglesias</span></p>
            <p class="dateline"><a href="posts/2014/10/15/continuous-integration-in-python-5-report-test-coverage-using-coveralls/" rel="bookmark"><time class="post-date published dt-published" datetime="2014-10-15T23:11:19+11:00" title="2014-10-15 23:11">2014-10-15 23:11</time></a></p>
                <p class="commentline">
        
    <a href="posts/2014/10/15/continuous-integration-in-python-5-report-test-coverage-using-coveralls/#disqus_thread" data-disqus-identifier="cache/posts/2014/10/15/continuous-integration-in-python-5-report-test-coverage-using-coveralls.html">Comments</a>


        </p>
</div>
    </header><div class="e-content entry-content">
    <div>
<p></p>
<p>In this series of posts, we've covered:</p>
<p></p>
<ul>
<li><a href="http://ilovesymposia.com/2014/10/01/continuous-integration-0-automated-tests-with-pytest/">Automated testing with pytest</a></li>
<li><a href="http://ilovesymposia.com/2014/10/02/continuous-integration-1-test-coverage/">Measuring test coverage</a></li>
<li><a href="http://ilovesymposia.com/2014/10/13/continuous-integration-in-python-3-set-up-your-test-configuration-files/">Setting up test configuration files</a></li>
<li>Using Travis-CI to <a href="http://ilovesymposia.com/2014/10/15/continuous-integration-in-python-4-set-up-travis-ci/">run your tests automatically with each push</a>
</li>
</ul>
<p>Today I will show you how to continuously check your test coverage using <a href="https://coveralls.io">Coveralls</a>.</p>
<p>Travis runs whatever commands you tell it to run in your <code>.travis.yml</code> file. Normally, that's just installing your program and its requirements, and running your tests. If you wanted instead to launch some nuclear missiles, you could do that. (Assuming you were happy to put the launch keys in a public git repository... =P)</p>
<p>The <a href="https://coveralls.io">Coveralls</a> service, once again free for open-source repositories, takes advantage of this: you just need to install an extra piece of software from PyPI, and run it after your tests have passed. Do so by adding the line <code>pip install coveralls</code> to your <code>before_install</code> section, and just the <code>coveralls</code> command to a new <code>after_success</code> section:</p>
<p>```https://coveralls.io
; then
          coveralls;
      fi</p>
<p>```</p>
<p>This is the approach taken by the scikit-image project.)</p>
<p><strong>Update:</strong> <a href="http://ilovesymposia.com/2014/10/17/continuous-integration-in-python-6-show-off-your-work/">Volume 6: Badge your repo to show off your work</a></p>
</div>
    </div>
    </article>
</div>

        <nav class="postindexpager"><ul class="pager">
<li class="previous">
                <a href="index-6.html" rel="prev">Newer posts</a>
            </li>
            <li class="next">
                <a href="index-4.html" rel="next">Older posts</a>
            </li>
        </ul></nav><script>var disqus_shortname="nikolademo";(function(){var a=document.createElement("script");a.async=true;a.src="https://"+disqus_shortname+".disqus.com/count.js";(document.getElementsByTagName("head")[0]||document.getElementsByTagName("body")[0]).appendChild(a)}());</script><footer id="footer"><p>Contents © 2019         <a href="mailto:jni.soma@fastmail.com">Juan Nunez-Iglesias</a> - Powered by         <a href="https://getnikola.com" rel="nofollow">Nikola</a>         </p>
            
        </footer>
</div>
    </div>
    <label for="sidebar-checkbox" class="sidebar-toggle"></label>
    
    
    
            <script src="assets/js/all-nocdn.js"></script><!-- fancy dates --><script>
    moment.locale("en");
    fancydates(0, "YYYY-MM-DD HH:mm");
    </script><!-- end fancy dates -->
</body>
</html>
