<!DOCTYPE html>
<html prefix="" lang="en">
<head>
<meta charset="utf-8">
<meta name="description" content="Science &amp; Tech for the Small Fry">
<meta name="viewport" content="width=device-width">
<title>I Love Symposia! (old posts, page 4) | I Love Symposia!</title>
<link href="assets/css/all.css" rel="stylesheet" type="text/css">
<link rel="stylesheet" href="//fonts.googleapis.com/css?family=PT+Serif:400,400italic,700%7CPT+Sans:400">
<link rel="alternate" type="application/rss+xml" title="RSS" href="rss.xml">
<link rel="canonical" href="https://ilovesymposia.com/index-4.html">
<link rel="prev" href="index-5.html" type="text/html">
<link rel="next" href="index-3.html" type="text/html">
<script type="text/x-mathjax-config">
MathJax.Hub.Config({
    tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true
    },
    displayAlign: 'center', // Change this to 'left' if you want left-aligned equations.
    "HTML-CSS": {
        styles: {'.MathJax_Display': {"margin": 0}}
    }
});
</script><!--[if lt IE 9]><script src="assets/js/html5.js"></script><![endif]-->
</head>
<body>
    <a href="#content" class="sr-only sr-only-focusable">Skip to main content</a>
    <!-- Target for toggling the sidebar `.sidebar-checkbox` is for regular
            styles, `#sidebar-checkbox` for behavior. -->
    <input type="checkbox" class="sidebar-checkbox" id="sidebar-checkbox"><!-- Toggleable sidebar --><div class="sidebar" id="sidebar">
        <div class="sidebar-item">
            <p>
            A blog about Python, science, and the odd random thing I care enough about.
            </p>
            <p>
            Theme modified from <a href="https://themes.getnikola.com/v7/lanyon/">Lanyon</a>.
            Created by <a href="https://twitter.com/mdo" target="_blank">@mdo</a> for Jekyll,
            ported to <a href="https://getnikola.com">Nikola</a> by
            <a href="https://twitter.com/ralsina" target="_blank">@ralsina</a>.
            </p>
        </div>
        
    <nav id="menu" role="navigation" class="sidebar-nav"><a class="sidebar-nav-item" href="archive.html">Archives</a>
        <a class="sidebar-nav-item" href="categories/index.html">Tags</a>
        <a class="sidebar-nav-item" href="rss.xml">RSS feed</a>
    
    
    </nav>
</div>

    <!-- Wrap is the content to shift when toggling the sidebar. We wrap the
         content to avoid any CSS collisions with our real content. -->
    <div class="wrap">
      <div class="masthead">
        <div class="container">
          
    <h3 id="brand" class="masthead-title">
      <a href="https://ilovesymposia.com/" title="I Love Symposia!" rel="home">I Love Symposia!</a>
    </h3>

        </div>
      </div>

      <div class="container content" id="content">
        

<div class="posts">
    <article class="post h-entry post-text"><header><h1 class="post-title p-name"><a href="2014/10/15/continuous-integration-in-python-4-set-up-travis-ci/" class="u-url">Continuous integration in Python, 4: set up Travis-CI</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn">Juan Nunez-Iglesias</span></p>
            <p class="dateline"><a href="2014/10/15/continuous-integration-in-python-4-set-up-travis-ci/" rel="bookmark"><time class="post-date published dt-published" datetime="2014-10-15T04:00:59+11:00" title="2014-10-15 04:00">2014-10-15 04:00</time></a></p>
                <p class="commentline">
        
    <a href="2014/10/15/continuous-integration-in-python-4-set-up-travis-ci/#disqus_thread" data-disqus-identifier="cache/posts/2014/10/15/continuous-integration-in-python-4-set-up-travis-ci.html">Comments</a>


        </p>
</div>
    </header><div class="e-content entry-content">
    <div>
<ul>
<li><a href="http://ilovesymposia.com/2014/10/01/continuous-integration-0-automated-tests-with-pytest/">Automated testing with pytest</a></li>
<li><a href="http://ilovesymposia.com/2014/10/02/continuous-integration-1-test-coverage/">Measuring test coverage</a></li>
<li><a href="http://ilovesymposia.com/2014/10/13/continuous-integration-in-python-3-set-up-your-test-configuration-files/">Setting up test configuration files</a></li>
</ul>
<h3>Introduction to Travis-CI</h3>
<p>Once you've set up your tests locally, it does you no good if you don't remember to run them! <a href="http://travis-ci.org/">Travis-CI</a> makes this seamless, because it will check out your code and run you tests for <em>each and every commit you push to GitHub!</em> (This is even more important when you are receiving pull requests on GitHub: the tests will be run online, without you having to individually check out each PR and run the tests on your machine!)
.. has_math: no
.. status: published
.. wp-status: publish
--&gt;</p>
<p></p>
<p>This is the fourth post in my Continuous integration (CI) in Python series, and the one that puts the "continuous" in CI! For the introductory three posts, see:</p>
<p></p>
<ul>
<li><a href="http://ilovesymposia.com/2014/10/01/continuous-integration-0-automated-tests-with-pytest/">Automated testing with pytest</a></li>
<li><a href="http://ilovesymposia.com/2014/10/02/continuous-integration-1-test-coverage/">Measuring test coverage</a></li>
<li><a href="http://ilovesymposia.com/2014/10/13/continuous-integration-in-python-3-set-up-your-test-configuration-files/">Setting up test configuration files</a></li>
</ul>
<h3>Introduction to Travis-CI</h3>

<p>Once you've set up your tests locally, it does you no good if you don't remember to run them! <a href="http://travis-ci.org/">Travis-CI</a> makes this seamless, because it will check out your code and run you tests for <em>each and every commit you push to GitHub!</em> (This is even more important when you are receiving pull requests on GitHub: the tests will be run online, without you having to individually check out each PR and run the tests on your machine!)</p>
<p>This is what continuous integration is all about. Once upon a time, the common practice was to pile on new features on a codebase. Then, come release time, there would be a feature freeze, and some time would be spent cleaning up code and removing bugs. In continuous integration, instead, no new feature is allowed into the codebase until it is bug free, as demonstrated by the test suite.</p>
<h3>What to do</h3>

<p>You need to first add a <code>.travis.yml</code> file to the root of your project. This tells Travis how to install your program's dependencies, install your program, and run the tests. Here's an example file to run tests and coverage on our <code>maths.py</code> sample project:</p>
<pre class="code literal-block"><span></span><span class="k">[code lang=text]</span>
<span class="na">language: python</span>
<span class="na">python:</span>
    <span class="na">- "2.7"</span>
    <span class="na">- "3.4"</span>
<span class="na">before_install:</span>
    <span class="na">- pip install pytest pytest-cov</span>
<span class="na">script:</span>
    <span class="na">- py.test</span>
<span class="k">[/code]</span>
</pre>


<p>Pretty simple: tell Travis the language of your project, the Python version (you can specify multiple versions, one per line, to test on multiple Python versions!), how to install test dependencies. Finally, the command to run your tests. (This can be anything, not just pytest or another testing framework, as long as a shell exit status of 0 means "success" and anything else means "failure".)</p>
<p>You can read more about the syntax and options for your <code>.travis.yml</code> in the <a href="http://docs.travis-ci.com/user/build-configuration/">Travis documentation</a>. There are other sections you can add, such as "virtualenv" to set up Python virtual environments, "install" to add compilation and other installation steps for your library, before testing, and "after_success" to enable e.g. custom notifications. (We will use this last one in the next post.)</p>
<p>Once you have created <code>.travis.yml</code> for your project, you need to turn it on for your repo. <strike>This is, currently, a fairly painful process, and I can't wait for the day that it's enabled by default on GitHub.</strike> [<strong>Update: see below</strong>] In the meantime though:</p>
<ul>
<li><strike>Go to your project <a href="2014/10/skitch-3.png">settings</a> page on GitHub.</strike></li>
<li><strike>Click on <a href="2014/10/skitch-2.png">webhooks &amp; services</a>.</strike></li>
<li><strike>Click on <a href="2014/10/skitch.png">add service</a>, searching for "Travis".</strike></li>
<li><strike>You'll be taken to the GitHub <a href="2014/10/skitch-7.png">Travis configuration page</a>. Mine was pre-populated with the Travis token, but you might have to click on the link to your <a href="https://travis-ci.org/profile">Travis profile</a>, and click on the "Profile" tab to retrieve your token.</strike></li>
<li>In the <a href="https://travis-ci.org/profile">Travis profile page</a>, in the "Repositories" tab, click <a href="2014/10/skitch-5.png">Sync now</a> to get your list of repos, and flick the switch for your current repo.</li>
</ul>
<p><i>[<strong>Update 2014-10-28:</strong> Thanks to <a href="https://twitter.com/hugovk">@hugovk</a> for pointing out that the first four points above can be skipped. It turns out that when you first log in to Travis-CI using your GitHub account, you give them write access to your webhooks. So, when you add a repo from their end, they go ahead and add themselves on the GitHub end! Boom. Way easier.]</i></p>
<p>https://twitter.com/hugovk/status/526798595442618368</p>
<p>Voil√†! Every push and pull-request to your repository will trigger a job on Travis's servers, building your dependencies and your software, running your tests, and emailing you if anything went wrong! Amazingly, this is completely free for open source projects, so you really have no excuse for not using it!</p>
<p><img src="2014/10/screen-shot-2014-10-15-at-9-55-05-pm.png" alt="Travis build page"></p>
<p>Follow this blog to learn how to continuously check test coverage using <a href="https://coveralls.io">Coveralls</a>, coming in the next post!</p>
<p><strong>Update:</strong> <a href="http://ilovesymposia.com/2014/10/15/continuous-integration-in-python-5-report-test-coverage-using-coveralls/">Volume 5: turn on Coveralls</a></p>
</div>
    </div>
    </article><article class="post h-entry post-text"><header><h1 class="post-title p-name"><a href="2014/10/13/continuous-integration-in-python-3-set-up-your-test-configuration-files/" class="u-url">Continuous integration in Python 3: set up your test configuration files</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn">Juan Nunez-Iglesias</span></p>
            <p class="dateline"><a href="2014/10/13/continuous-integration-in-python-3-set-up-your-test-configuration-files/" rel="bookmark"><time class="post-date published dt-published" datetime="2014-10-13T19:57:20+11:00" title="2014-10-13 19:57">2014-10-13 19:57</time></a></p>
                <p class="commentline">
        
    <a href="2014/10/13/continuous-integration-in-python-3-set-up-your-test-configuration-files/#disqus_thread" data-disqus-identifier="cache/posts/2014/10/13/continuous-integration-in-python-3-set-up-your-test-configuration-files.html">Comments</a>


        </p>
</div>
    </header><div class="e-content entry-content">
    <div>
<p></p>
<p>This is the third post in my series on continuous integration (CI) in Python. For the first two posts, see <a href="http://ilovesymposia.com/2014/10/01/continuous-integration-0-automated-tests-with-pytest/">1: automated tests with pytest</a>, and <a href="http://ilovesymposia.com/2014/10/02/continuous-integration-1-test-coverage/">2: measuring test coverage</a>.</p>
<p>By now, you've got a bunch of tests and doctests set up in your project, which you run with the command:</p>
<p>```http://en.wikipedia.org/wiki/INI_file</p>
<p>omit = *script.py</p>
<p>```</p>
<p>Once you've done this, you can invoke your tests with a simple, undecorated call to just <code>py.test</code>.</p>
<p>To find out more about your configuration options, see the pytest <a href="http://pytest.org/latest/customize.html">basic test configuration</a> page, and Ned Batchelder's excellent <a href="http://nedbatchelder.com/code/coverage/config.html"><code>.coveragerc</code> syntax guide</a>.</p>
<p>That's it for this entry of my CI series. Follow this blog for the next two entries, setting up Travis CI and Coveralls.</p>
<p><strong>Update:</strong> <a href="http://ilovesymposia.com/2014/10/15/continuous-integration-in-python-4-set-up-travis-ci/">Volume 4: Set up Travis-CI</a></p>
</div>
    </div>
    </article><article class="post h-entry post-text"><header><h1 class="post-title p-name"><a href="2014/10/02/continuous-integration-1-test-coverage/" class="u-url">Continuous integration in Python, Volume 2: measuring test coverage</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn">Juan Nunez-Iglesias</span></p>
            <p class="dateline"><a href="2014/10/02/continuous-integration-1-test-coverage/" rel="bookmark"><time class="post-date published dt-published" datetime="2014-10-02T01:04:04+10:00" title="2014-10-02 01:04">2014-10-02 01:04</time></a></p>
                <p class="commentline">
        
    <a href="2014/10/02/continuous-integration-1-test-coverage/#disqus_thread" data-disqus-identifier="cache/posts/2014/10/02/continuous-integration-1-test-coverage.html">Comments</a>


        </p>
</div>
    </header><div class="e-content entry-content">
    <div>
<p></p>
<p>(<strong>Edit</strong>: I initially thought it would be cute to number from 0. But it turns out it becomes rather obnoxious to relate English (first, second, ...) to 0-indexing. So this was formerly volume 1. But everything else remains the same.)</p>
<p>This is the second post in a series about setting up continuous integration for a Python project from scratch. For the first post, see <a href="http://ilovesymposia.com/2014/10/01/continuous-integration-0-automated-tests-with-pytest/">Automated tests with pytest</a>.</p>
<p>After you've written some test cases for a tiny project, it's easy to check what code you have automatically tested. For even moderately big projects, you will need tools that automatically check what parts of your code are actually tested. The proportion of lines of code that are run at least once during your tests is called your <em>test coverage</em>.</p>
<p>For the same reasons that testing is important, measuring coverage is important. Pytest can measure coverage for you with the coverage plugin, found in the <a href="https://pypi.python.org/pypi/pytest-cov">pytest-cov</a> package. Once you've installed the extension, a test coverage measurement is just a command-line option away:</p>
<pre class="code literal-block"><span></span><span class="k">[code lang=text]</span>
 <span class="na">~/projects/maths $ py.test --doctest-modules --cov .</span>
<span class="o">=</span><span class="s">============================ test session starts ==============================</span>
<span class="na">platform darwin -- Python 2.7.8 -- py-1.4.25 -- pytest-2.6.3</span>
<span class="na">plugins: cov</span>
<span class="na">collected 2 items </span>

<span class="na">maths.py .</span>
<span class="na">test_maths.py .</span>
<span class="na">--------------- coverage: platform darwin, python 2.7.8-final-0 ----------------</span>
<span class="na">Name         Stmts   Miss  Cover</span>
<span class="na">--------------------------------</span>
<span class="na">maths            2      0   100%</span>
<span class="na">test_maths       4      0   100%</span>
<span class="na">--------------------------------</span>
<span class="na">TOTAL            6      0   100%</span>

<span class="o">=</span><span class="s">========================== 2 passed in 0.07 seconds ===========================</span>
<span class="k">[/code]</span>
</pre>


<p>(The <code>--cov</code> takes a directory as input, which I find obnoxious, given that py.test so naturally defaults to the current directory. But it is what it is.)</p>
<p>Now, if I add a function without a test, I'll see my coverage drop:</p>
<pre class="code literal-block"><span></span><span class="k">[code lang=text]</span>
<span class="na">def sqrt(x):</span>
    <span class="na">"""Return the square root of `x`."""</span>
    <span class="na">return x * 0.5</span>
<span class="k">[/code]</span>
</pre>


<p>(The typo is intentional.)</p>
<pre class="code literal-block"><span></span><span class="k">[code lang=text]</span>
<span class="na">--------------- coverage: platform darwin, python 2.7.8-final-0 ----------------</span>
<span class="na">Name         Stmts   Miss  Cover</span>
<span class="na">--------------------------------</span>
<span class="na">maths            4      1    75%</span>
<span class="na">test_maths       4      0   100%</span>
<span class="na">--------------------------------</span>
<span class="na">TOTAL            8      1    88%</span>
<span class="k">[/code]</span>
</pre>


<p>With one more option, <code>--cov-report term-missing</code>, I can see which lines I <em>haven't</em> covered, so I can try to design tests specifically for that code:</p>
<pre class="code literal-block"><span></span><span class="k">[code lang=text]</span>
<span class="na">--------------- coverage: platform darwin, python 2.7.8-final-0 ----------------</span>
<span class="na">Name         Stmts   Miss  Cover   Missing</span>
<span class="na">------------------------------------------</span>
<span class="na">maths            4      1    75%   24</span>
<span class="na">test_maths       4      0   100%   </span>
<span class="na">------------------------------------------</span>
<span class="na">TOTAL            8      1    88%   </span>
<span class="k">[/code]</span>
</pre>


<p>Do note that 100% coverage does not ensure correctness. For example, suppose I test my <code>sqrt</code> function like so:</p>
<pre class="code literal-block"><span></span><span class="k">[code lang=python]</span>
<span class="na">def sqrt(x):</span>
    <span class="na">"""Return the square root of `x`.</span>

    <span class="na">Examples</span>
    <span class="na">--------</span>
    <span class="na">&amp;gt;&amp;gt;&amp;gt; sqrt(4.0)</span>
    <span class="na">2.0</span>
    <span class="na">"""</span>
    <span class="na">return x * 0.5</span>
<span class="k">[/code]</span>
</pre>


<p>Even though my test is correct, and I now have 100% test coverage, I haven't detected my mistake. Oops!</p>
<p>But, keeping that caveat in mind, full test coverage <em>is</em> a wonderful thing, and if you don't test something, you're guaranteed not to catch errors. Further, my example above is quite contrived, and in most situations full test coverage <em>will</em> spot most errors.</p>
<p>That's it for part 2. Tune in next time to learn how to turn on <a href="https://travis-ci.org">Travis</a> continuous integration for your GitHub projects!</p>
<p><strong>Update:</strong> <a href="http://ilovesymposia.com/2014/10/13/continuous-integration-in-python-3-set-up-your-test-configuration-files/">Volume 3: set up config files.</a></p>
</div>
    </div>
    </article><article class="post h-entry post-text"><header><h1 class="post-title p-name"><a href="2014/10/01/continuous-integration-0-automated-tests-with-pytest/" class="u-url">Continuous integration in Python, Volume 1: automated tests with pytest</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn">Juan Nunez-Iglesias</span></p>
            <p class="dateline"><a href="2014/10/01/continuous-integration-0-automated-tests-with-pytest/" rel="bookmark"><time class="post-date published dt-published" datetime="2014-10-01T01:40:41+10:00" title="2014-10-01 01:40">2014-10-01 01:40</time></a></p>
                <p class="commentline">
        
    <a href="2014/10/01/continuous-integration-0-automated-tests-with-pytest/#disqus_thread" data-disqus-identifier="cache/posts/2014/10/01/continuous-integration-0-automated-tests-with-pytest.html">Comments</a>


        </p>
</div>
    </header><div class="e-content entry-content">
    <div>
<p>Without further ado, here goes the first post: how to set up automated testing for your Python project using <a href="http://pytest.org/latest/">pytest</a>.</p>
<p>.. has_math: no
.. status: published
.. wp-status: publish
--&gt;</p>
<p></p>
<p>(<strong>Edit</strong>: I initially thought it would be cute to number from 0. But it turns out it becomes rather obnoxious to relate English (first, second, ‚Ä¶) to 0-indexing. So this was formerly volume 0. But everything else remains the same.)</p>
<p>I just finished the process of setting up continuous integration <em>from scratch</em> for one of my projects, <a href="https://github.com/jni/cellom2tif">cellom2tif</a>, a simple image file converter/liberator. I thought I would write a blog post about that process, but it has slowly mutated into a hefty document that I thought would work better as a series. I'll cover automated testing, test coverage, and how to get these to run automatically for your project with <a href="https://travis-ci.org">Travis-CI</a> and <a href="https://coveralls.io">Coveralls</a>.</p>
<p>Without further ado, here goes the first post: how to set up automated testing for your Python project using <a href="http://pytest.org/latest/">pytest</a>.</p>
<p></p>
<h3>Automated tests, and why you need them</h3>
<p>Software engineering is hard, and it's incredibly easy to mess things up, so you should write <em>tests</em> for all your functions, which ensure that nothing obviously stupid is going wrong. Tests can take a lot of different forms, but here's a really basic example. Suppose this is a function in your file, <code>maths.py</code>:</p>
<p>```http://ilovesymposia.com/2014/01/09/best-practices-addendum-find-and-follow-the-conventions-of-your-programming-community/</p>
<p>~/projects/maths $ py.test --doctest-modules
============================= test session starts ==============================
platform darwin -- Python 2.7.8 -- py-1.4.20 -- pytest-2.5.2
collected 2 items</p>
<p>maths.py .
test_maths.py .</p>
<p>=========================== 2 passed in 0.06 seconds ===========================</p>
<p>```</p>
<h3>Test-driven development</h3>

<p>That was easy! And yet most people, my past self included, neglect tests, thinking they'll do them eventually, when the software is ready. This is backwards. You've probably heard the phrase "Test-driven development (TDD)"; this is what they're talking about: writing your tests <em>before</em> you've written the functionality to pass them. It might initially <em>seem</em> like wasted effort, like you're not making progress in what you actually want to do, which is write your software. But it's not:</p>
<p>https://twitter.com/zspencer/status/514447236239859712</p>
<p>By spending a bit of extra effort to prevent bugs down the road, you will get to where you want to go faster.</p>
<p>That's it for volume 1! Watch out for the next post: ensuring your tests are thorough by measuring <em>test coverage</em>.</p>
<p><strong>Update:</strong> <a href="http://ilovesymposia.com/2014/10/02/continuous-integration-1-test-coverage/">Volume 2: measuring test coverage</a></p>
</div>
    </div>
    </article><article class="post h-entry post-text"><header><h1 class="post-title p-name"><a href="2014/09/22/use-pmset-to-restore-your-macs-instant-wake-after-upgrading-to-yosemite/" class="u-url">Use pmset to restore your Mac's instant-wake after upgrading to Yosemite</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn">Juan Nunez-Iglesias</span></p>
            <p class="dateline"><a href="2014/09/22/use-pmset-to-restore-your-macs-instant-wake-after-upgrading-to-yosemite/" rel="bookmark"><time class="post-date published dt-published" datetime="2014-09-22T04:14:14+10:00" title="2014-09-22 04:14">2014-09-22 04:14</time></a></p>
                <p class="commentline">
        
    <a href="2014/09/22/use-pmset-to-restore-your-macs-instant-wake-after-upgrading-to-yosemite/#disqus_thread" data-disqus-identifier="cache/posts/2014/09/22/use-pmset-to-restore-your-macs-instant-wake-after-upgrading-to-yosemite.html">Comments</a>


        </p>
</div>
    </header><div class="e-content entry-content">
    <div>
<p></p>
<p>I just upgraded to Mac OS X Yosemite Beta 3, and sure enough, Apple has <a href="http://ilovesymposia.com/2013/11/05/speed-up-your-macs-wake-up-time-using-pmset-do-it-again-after-upgrading-to-mavericks/">once again</a> broken my time-to-deep-sleep setting. After I upgraded, I noticed that my computer took a few seconds to wake from sleep, rather than the usual instant-on that defined OS X for so long. I checked <code>pmset</code>, Mac's power management command-line utility:</p>
<p>```http://ilovesymposia.com/2013/11/05/speed-up-your-macs-wake-up-time-using-pmset-do-it-again-after-upgrading-to-mavericks/</p>
<p>sudo pmset -a standbydelay 86400</p>
<p>```</p>
<p>Check in for a new post come next upgrade! (Mammoth?)</p>
</div>
    </div>
    </article><article class="post h-entry post-text"><header><h1 class="post-title p-name"><a href="2014/08/10/read-microscopy-images-to-numpy-arrays-with-python-bioformats/" class="u-url">Read microscopy images to numpy arrays with python-bioformats</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn">Juan Nunez-Iglesias</span></p>
            <p class="dateline"><a href="2014/08/10/read-microscopy-images-to-numpy-arrays-with-python-bioformats/" rel="bookmark"><time class="post-date published dt-published" datetime="2014-08-10T00:43:36+10:00" title="2014-08-10 00:43">2014-08-10 00:43</time></a></p>
                <p class="commentline">
        
    <a href="2014/08/10/read-microscopy-images-to-numpy-arrays-with-python-bioformats/#disqus_thread" data-disqus-identifier="cache/posts/2014/08/10/read-microscopy-images-to-numpy-arrays-with-python-bioformats.html">Comments</a>


        </p>
</div>
    </header><div class="e-content entry-content">
    <div>
<p></p>
<p>The python-bioformats library lets you seamlessly read microscopy images into numpy arrays from pretty much any file format.</p>
<p>I <a href="http://ilovesymposia.com/2014/01/15/fiji-jython">recently explained</a> how to use Fiji's Jython interpreter to open BioFormats images, do some processing, and save the result to a standard format such as TIFF. Since then, the
<a href="http://cellprofiler.org">CellProfiler team</a> has released an incredible tool, the <a href="https://github.com/cellprofiler/python-bioformats">python-bioformats</a> library. It uses the Java Native Interface (JNI) to access the Java BioFormats library and give you back a numpy array within Python. In other words, it's magic.</p>
<p>Some of the stuff I was doing in the Jython interpreter was not going to fly for a 400GB image file produced by Leica (namely, setting the flag <code>setOpenAllSeries(True)</code>). This file contained 3D images of multiple zebrafish embryos, obtained every 30 minutes for three days. I needed to process each image sequentially.</p>
<p>The first problem was that even reading the metadata from the file <a href="https://github.com/CellProfiler/python-bioformats/issues/8">resulted in a Java out-of-memory error</a>! With the help of Lee Kamentsky, one of the creators of python-bioformats, I figured out that Java allocates a maximum memory footprint of just 256MB. With the raw metadata string occupying 27MB, this was not enough to contain the full structure of the parsed metadata tree. The solution was simply to set a much larger maximum memory allocation to the JVM:</p>
<p>```https://github.com/jni/lesion/blob/c3223687d35a7f81da7305e1e041f9c5a53104b1/lesion/lifio.py#L79
, cmap=cm.gray)
plt.show()</p>
<p>```</p>
<p><img src="http://ilovesymposia.files.wordpress.com/2014/08/embryo-image.png" alt="2D slice from 5D volume"></p>
<p>Boom. Using Python BioFormats, I've read in a small(ish) part of a quasi-terabyte image file into a numpy array, ready for further processing.</p>
<p>Note: the dimension order here is time, z, y, x, channel, or TZYXC, which I <em>think</em> is the most efficient way to read these files in. My <a href="https://github.com/jni/lesion/blob/c3223687d35a7f81da7305e1e041f9c5a53104b1/lesion/lifio.py#L181">wrapper</a> allows arbitrary dimension order, so it'll be good to use it to figure out the fastest way to iterate through the volume.</p>
<p>In my case, I'm looking to extract statistics using scikit-image's <a href="https://github.com/scikit-image/scikit-image/blob/master/skimage/measure/profile.py"><code>profile_line</code></a> function, and plot their evolution over time. Here's the min/max intensity <a href="https://github.com/jni/lesion/blob/c3223687d35a7f81da7305e1e041f9c5a53104b1/lesion/trace.py">profile</a> along the embryo for a sample stack:</p>
<p><img src="http://ilovesymposia.files.wordpress.com/2014/08/position02.png" alt="min/max vs time"></p>
<p>I still need to clean up the code, in particular to detect bad images (no prizes for guessing which timepoint was bad here), but my point for now is that, thanks to Python BioFormats, doing your entire bioimage analysis in Python just got a heck of a lot easier.</p>
</div>
    </div>
    </article><article class="post h-entry post-text"><header><h1 class="post-title p-name"><a href="2014/07/17/scipy-2014-recap/" class="u-url">SciPy 2014: an extremely useful conference with a diversity problem</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn">Juan Nunez-Iglesias</span></p>
            <p class="dateline"><a href="2014/07/17/scipy-2014-recap/" rel="bookmark"><time class="post-date published dt-published" datetime="2014-07-17T18:48:56+10:00" title="2014-07-17 18:48">2014-07-17 18:48</time></a></p>
                <p class="commentline">
        
    <a href="2014/07/17/scipy-2014-recap/#disqus_thread" data-disqus-identifier="cache/posts/2014/07/17/scipy-2014-recap.html">Comments</a>


        </p>
</div>
    </header><div class="e-content entry-content">
    <div>
<h3>About SciPy in general</h3>
<p>Since my first SciPy in 2012, I've decided to prioritise programming conferences over scientific ones, because the value I get for my time is just that much higher. At a scientific conference, I certainly become aware of way-cool stuff going on in other research labs in my area. Once I get home, however, I go back to whatever I was doing. In contrast, at programming conferences, I become aware of new tools and practices that <em>change the way I do science.</em> In his <a href="https://www.youtube.com/watch?v=1e26rp6qPbA">keynote</a> this year, Greg Wilson said of Software Carpentry, "We save researchers a day a week for the rest of their careers." I feel the same way about SciPy in general.</p>
<p>In the 2012 sprints, I learned about <a href="https://help.github.com/articles/using-pull-requests">GitHub Pull Requests</a> and code review, the lingua franca of open source development today. I can't express how useful that's been. I also started my ongoing collaboration with the <a href="http://scikit-image.org/">scikit-image</a> project, which has enabled my research to reach far more users than I ever could have achieved on my own.</p>
<p>No scientific conference I've been to has had such an impact on my science output, nor can I imagine one doing so.
.. has_math: no
.. status: published
.. wp-status: publish
--&gt;</p>
<p></p>
<p>I just got back home from the <a href="https://conference.scipy.org/scipy2014/">SciPy 2014</a> conference in Austin, TX. Here are my thoughts after this year's conference.</p>
<p></p>
<h3>About SciPy in general</h3>
<p>Since my first SciPy in 2012, I've decided to prioritise programming conferences over scientific ones, because the value I get for my time is just that much higher. At a scientific conference, I certainly become aware of way-cool stuff going on in other research labs in my area. Once I get home, however, I go back to whatever I was doing. In contrast, at programming conferences, I become aware of new tools and practices that <em>change the way I do science.</em> In his <a href="https://www.youtube.com/watch?v=1e26rp6qPbA">keynote</a> this year, Greg Wilson said of Software Carpentry, "We save researchers a day a week for the rest of their careers." I feel the same way about SciPy in general.</p>
<p>In the 2012 sprints, I learned about <a href="https://help.github.com/articles/using-pull-requests">GitHub Pull Requests</a> and code review, the lingua franca of open source development today. I can't express how useful that's been. I also started my ongoing collaboration with the <a href="http://scikit-image.org/">scikit-image</a> project, which has enabled my research to reach far more users than I ever could have achieved on my own.</p>
<p>No scientific conference I've been to has had such an impact on my science output, nor can I imagine one doing so.</p>
<h3>This year's highlights</h3>

<p>This year was no different. Without further ado, here are my top hits from this year's conference:</p>
<ul>
<li>Michael Droettboom <a href="https://www.youtube.com/watch?v=OsxJ5O6h8s0">talked</a> about his continuous benchmarking project, <a href="https://spacetelescope.github.io/asv/">Airspeed Velocity</a>. It is hilariously named and incredibly useful. <code>asv</code> checks out code from your Git repo at regular intervals and runs benchmarks (which you define), and plots your code's performance over time. It's an incredible guard against feature creep slowing down your code base.</li>
<li>IPython recently unveiled their modal version 2.0 interface, sending vimmers worldwide rejoicing. But a few key bindings are just <em>wrong</em> from a vim perspective. Most egregiously, <code>i</code>, which should enter edit mode, interrupts the kernel when pressed twice! That's just evil. Paul Ivanov goes all in with vim keybindings with his hilarious and life-changing <a href="https://www.youtube.com/watch?v=p9gnhmX1sPo">IPython vimception</a> talk. The title is more appropriate than I realised. Must-watch.</li>
<li>Dami√°n Avila <a href="https://www.youtube.com/watch?v=sZBKruEh0jI">revealed</a> (heh) his live IPython presentations with Reveal.js, forever changing how Python programmers present their work. I was actually aware of this before the conference and used it in <a href="https://www.youtube.com/watch?v=fn8F_NerTug">my own talk</a>, but you should definitely watch his talk and get the extension from his <a href="https://github.com/damianavila/live_reveal">repo</a>.</li>
<li>Min RK gave an excellent tutorial on IPython parallel (<a href="https://github.com/minrk/IPython-parallel-tutorial">repo</a>, videos <a href="https://www.youtube.com/watch?v=y4hgalfhc1Y">1</a>, <a href="https://www.youtube.com/watch?v=-9ijnHPCYhY">2</a>, <a href="https://www.youtube.com/watch?v=U5mhpKkIx2Y">3</a>). It's remarkable what the IPython team have achieved thanks to their decoupling of the interactive shell and the computation "kernel". (I still hate that word.)</li>
<li>Brian Granger and Jon Frederic gave an excellent <a href="https://www.youtube.com/watch?v=aIXED26Wppg">tutorial</a> on IPython interactive widgets (notebook <a href="https://github.com/ipython/ipython-in-depth">here</a>). They provide a simple and fast way to interactively explore your data. I've already started using these on my own problems.</li>
<li>Aaron Meurer gave the <a href="https://www.youtube.com/watch?v=UaIvrDWrIWM">best introduction</a> to the Python packaging problem that I've ever seen, and how Continuum's conda project solves it. I think we still need an in-depth tutorial on how package <em>distributors</em> should use conda, but for users, conda is just awesome, and this talk explains why.</li>
<li>Matt Rocklin has a gift for crystal clear speaking, despite his incredible speed, and it was on full display in his (and Mark Wiebe's) talk on <a href="https://www.youtube.com/watch?v=9HPR-1PdZUk">Blaze</a>, Continuum's new array abstraction library. I'm not sure I'll be using Blaze immediately but it's certainly on my radar now!</li>
<li>Lightning talks are always fun: days <a href="https://www.youtube.com/watch?v=JDrhn0-r9Eg">1</a>, <a href="https://www.youtube.com/watch?v=SMyto7WHiNs">2</a>, <a href="https://www.youtube.com/watch?v=ln4nE_EVDCg">3</a>. Watch out for Fernando P√©rez's announcement of Project Jupyter, the evolution of the IPython notebook, and for Damon McDougall's riveting history of waffles. (You think I'm joking.)</li>
</ul>
<p>Apologies if I've missed anyone: with three tracks, an added one with the World Cup matches ;) , and my own talk preparations, "overwhelming" does not begin to describe the conference! I will second Aaron Meurer's <a href="https://asmeurer.github.io/blog/posts/scipy-2014/">assertion</a> that there were no bad talks. Which brings us to...</p>
<h3>On my to-watch</h3>

<p>Jake Vanderplas recently wrote a series of blog posts (last one <a href="http://jakevdp.github.io/blog/2014/06/14/frequentism-and-bayesianism-4-bayesian-in-python/">here</a>, with links to earlier posts) comparing frequentist and Bayesian approaches to statistical analysis, in which he makes a compelling argument that we should all relegate frequentism to the dustbin of history. As such, I intend to go over Chris Fonnesbeck's <a href="https://www.youtube.com/watch?v=vOBB_ycQ0RA">tutorial</a> (<a href="https://www.youtube.com/watch?v=gFYPCdWB2-w">2</a>, <a href="https://www.youtube.com/watch?v=54sFjp7AvXM">3</a>) and <a href="https://www.youtube.com/watch?v=XbxIo7ScVzc">talk</a> about Bayesian analysis in Python using PyMC.</p>
<p>David Sanders also did a Julia <a href="https://www.youtube.com/watch?v=vWkgEddb4-A">tutorial</a> (<a href="https://www.youtube.com/watch?v=I3JH5Bg46yU">part 2</a>) that was scheduled at the same time as my own scikit-image tutorial, but I'm interested to see how the Julia ecosystem is progressing, so that should be a good place to start. (Although I'm still bitter that they went with 1-based indexing!)</p>
<p>The <a href="https://www.youtube.com/watch?v=EzX7MN_bzqg">reproducible science tutorial</a> (<a href="https://www.youtube.com/watch?v=HCyHn_by3N0">part 2</a>) generated quite a bit of buzz so I would be interested to go over that one as well.</p>
<p>For those interested in computing education or in geoscience, the conference had dedicated tracks for each of those, so you are bound to find something you like (not least, Lorena Barba's and Greg Wilson's keynotes). Have a look at the full listing of videos <a href="https://www.youtube.com/playlist?list=PLYx7XA2nY5GfuhCvStxgbynFNrxr3VFog">here</a>. These might be easier to navigate by looking at the <a href="https://conference.scipy.org/scipy2014/schedule/">conference schedule</a>.</p>
<h3>The SciPy experience</h3>

<p>I want to close this post with a few reflections on the conference itself.</p>
<p>SciPy is broken up into three "stages": two days of tutorials, three days of talks, and two days of sprints. Above, I covered the tutorials and talks, but to me, the sprints are what truly distinguish SciPy. The <a href="https://twitter.com/ellisonbg/status/487731404675899392">spirit of collaboration</a> is unparalleled, and an astonishing amount of value is generated in those two days, either in the shape of code, or in introducing newcomers to new projects and new ways to collaborate in programming.</p>
<p>My biggest regret of the conference was not giving a lightning talk urging people to come to the sprints. I repeatedly asked people whether they were coming to the sprints, and almost invariably the answer was that they didn't feel they were good enough to contribute. To reiterate my previous statements: (1) when I attended my first sprint in 2012, <em>I had never done a pull request</em>; (2) sprints are an excellent way to introduce newcomers to projects and to the pull request development model. All the buzz around the sprints was how welcoming all of the teams were, but I think there is a massive number of missed opportunities because this is not made obvious to attendees <em>before</em> the sprints.</p>
<p>Lastly, a few notes on diversity. During the conference, April Wright, a student in evolutionary biology at UT Austin, wrote a heartbreaking <a href="https://wrightaprilm.github.io/posts/lonely.html">blog post</a> about how excluded she felt from a conference where only 15% of attendees were women. That particular incident was joyfully resolved, with plenty of SciPyers reaching out to April and inviting her along to sprints and other events. But it highlighted just how poorly we are doing in terms of diversity. Andy Terrel, one of the conference organisers, <a href="https://twitter.com/aterrel/status/487281782530650112">pointed out</a> that 15% is much better than 2012's three (women, not percent!), but (a) that is still extremely low, and (b) I was horrified to read this because I was there in 2012... <em>And I did not notice that anything was wrong.</em> How can it be, in 2012, that it can seem <em>normal</em> to be at a professional conference and have effectively zero women around? It doesn't matter what one says about the background percentage of women in our industry and so on... Maybe SciPy is doing all it can about diversity. (Though I doubt it.) The point is that a scene like that should <em>feel</em> like one of those deserted cityscapes in post-apocalyptic movies. As long as it doesn't, as long as SciPy feels normal, we will continue to have diversity problems. I hope my fellow SciPyers look at these numbers, feel appalled as I have, and try to improve.</p>
<p>... And on cue, while I was writing this post, Andy Terrel wrote a great post of his own about this very topic:
http://andy.terrel.us/blog/2014/07/17/</p>
<p>I still consider SciPy a fantastic conference. Jonathan Eisen (<a href="https://twitter.com/phylogenomics">@phylogenomics</a>), whom I admire, would undoubtedly boycott it because of the problems outlined above, but I am heartened that the organising committee is taking this as a serious problem and trying hard fix it. I hope next time it is even better.</p>
</div>
    </div>
    </article><article class="post h-entry post-text"><header><h1 class="post-title p-name"><a href="2014/06/24/a-clever-use-of-scipys-ndimage-generic_filter-for-n-dimensional-image-processing/" class="u-url">A clever use of SciPy's ndimage.generic_filter for n-dimensional image processing</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn">Juan Nunez-Iglesias</span></p>
            <p class="dateline"><a href="2014/06/24/a-clever-use-of-scipys-ndimage-generic_filter-for-n-dimensional-image-processing/" rel="bookmark"><time class="post-date published dt-published" datetime="2014-06-24T01:48:07+10:00" title="2014-06-24 01:48">2014-06-24 01:48</time></a></p>
                <p class="commentline">
        
    <a href="2014/06/24/a-clever-use-of-scipys-ndimage-generic_filter-for-n-dimensional-image-processing/#disqus_thread" data-disqus-identifier="cache/posts/2014/06/24/a-clever-use-of-scipys-ndimage-generic_filter-for-n-dimensional-image-processing.html">Comments</a>


        </p>
</div>
    </header><div class="e-content entry-content">
    <div>
<p>Vighnesh is <a href="http://www.google-melange.com/gsoc/proposal/public/google/gsoc2014/vighneshbirodkar/5870670537818112">tasked</a> with implementing region adjacency graphs and graph based methods for image segmentation. He initially wrote specific functions for 2D and 3D images, and I suggested that he should merge them: either with n-dimensional code, or, at the very least, by making 2D a special case of 3D. He chose the former, and produced extremely elegant code. Three nested for loops and a large number of neighbour computations were replaced by a function call and a simple loop. Read on to find out how.</p>

<p>.. has_math: no
.. status: published
.. wp-status: publish
--&gt;</p>
<p></p>
<p>This year I am privileged to be a mentor in the <a href="https://www.google-melange.com/gsoc/homepage/google/gsoc2014">Google Summer of Code</a> for the <a href="http://scikit-image.org">scikit-image</a> project, as part of the Python Software Foundation organisation. Our student, <a href="https://github.com/vighneshbirodkar">Vighnesh Birodkar</a>, recently came up with a clever use of SciPy‚Äôs <a href="http://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.filters.generic_filter.html"><code>ndimage.generic_filter</code></a> that is certainly worth sharing widely.</p>
<p>Vighnesh is <a href="http://www.google-melange.com/gsoc/proposal/public/google/gsoc2014/vighneshbirodkar/5870670537818112">tasked</a> with implementing region adjacency graphs and graph based methods for image segmentation. He initially wrote specific functions for 2D and 3D images, and I suggested that he should merge them: either with n-dimensional code, or, at the very least, by making 2D a special case of 3D. He chose the former, and produced extremely elegant code. Three nested for loops and a large number of neighbour computations were replaced by a function call and a simple loop. Read on to find out how.</p>
<p>Iterating over an array of unknown dimension is not trivial a priori, but thankfully, someone else has already solved that problem: NumPy‚Äôs <a href="http://docs.scipy.org/doc/numpy/reference/generated/numpy.nditer.html"><code>nditer</code></a> and <a href="http://docs.scipy.org/doc/numpy/reference/generated/numpy.ndindex.html"><code>ndindex</code></a> functions allow one to efficiently iterate through every point of an n-dimensional array. However, that still leaves the problem of finding neighbors, to determine which regions are adjacent to each other. Again, this is not trivial to do in nD.</p>
<p>scipy.ndimage provides a suitable function, <code>generic_filter</code>. Typically, a filter is used to iterate a ‚Äúselector‚Äù (called a structuring element) over an array, compute some function of all the values covered by the structuring element, and replace the central value by the output of the function. For example, using the structuring element:</p>
<pre class="code literal-block"><span></span><span class="k">[code lang=python]</span>
<span class="na">fp</span> <span class="o">=</span> <span class="s">np.array([[0, 1, 0],</span>
<span class="s">               [1, 1, 1],</span>
<span class="s">               [0, 1, 0]], np.uint8)</span>
<span class="k">[/code]</span>
</pre>


<p>and the function <code>np.median</code> on a 2D image produces a median filter over a pixel‚Äôs immediate neighbors. That is,</p>
<pre class="code literal-block"><span></span><span class="k">[code lang=python]</span>
<span class="na">import functools</span>
<span class="na">median_filter</span> <span class="o">=</span> <span class="s">functools.partial(generic_filter,</span>
<span class="s">                                  function=np.median,</span>
<span class="s">                                  footprint=fp)</span>
<span class="k">[/code]</span>
</pre>


<p>Here, we don‚Äôt want to create an output array, but an output graph. What to do? As it turns out, Python‚Äôs pass-by-reference allowed Vighnesh to do this quite easily using the ‚Äúextra_arguments‚Äù keyword to <code>generic_filter</code>: we can write a filter function that receives the graph and updates it when two distinct values are adjacent! <code>generic_filter</code> passes all values covered by a structuring element as a flat array, in the array order of the structuring element. So Vighnesh wrote the following function:</p>
<pre class="code literal-block"><span></span><span class="k">[code lang=python]</span>
<span class="na">def _add_edge_filter(values, g):</span>
    <span class="na">"""Add an edge between first element in `values` and</span>
    <span class="na">all other elements of `values` in the graph `g`.</span>
    <span class="na">`values[0]` is expected to be the central value of</span>
    <span class="na">the footprint used.</span>

    <span class="na">Parameters</span>
    <span class="na">----------</span>
    <span class="na">values : array</span>
        <span class="na">The array to process.</span>
    <span class="na">g : RAG</span>
        <span class="na">The graph to add edges in.</span>

    <span class="na">Returns</span>
    <span class="na">-------</span>
    <span class="na">0.0 : float</span>
        <span class="na">Always returns 0.</span>

    <span class="na">"""</span>
    <span class="na">values</span> <span class="o">=</span> <span class="s">values.astype(int)</span>
<span class="s">    current = values[0]</span>
<span class="s">    for value in values[1:]:</span>
<span class="s">        g.add_edge(current, value)</span>
<span class="s">    return 0.0</span>
<span class="k">[/code]</span>
</pre>


<p>Then, using the footprint:</p>
<pre class="code literal-block"><span></span><span class="k">[code lang=python]</span>
<span class="na">fp</span> <span class="o">=</span> <span class="s">np.array([[0, 0, 0],</span>
<span class="s">               [0, 1, 1],</span>
<span class="s">               [0, 1, 0]], np.uint8)</span>
<span class="k">[/code]</span>
</pre>


<p>(or its n-dimensional analog), this filter is called as follows on <code>labels</code>, the image containing the region labels:</p>
<pre class="code literal-block"><span></span><span class="k">[code lang=python]</span>
<span class="na">filters.generic_filter(labels,</span>
                       <span class="na">function</span><span class="o">=</span><span class="s">_add_edge_filter,</span>
<span class="s">                       footprint=fp,</span>
<span class="s">                       mode='nearest',</span>
<span class="s">                       extra_arguments=(g,))</span>
<span class="k">[/code]</span>
</pre>


<p>This is a rather unconventional use of generic_filter, which is normally used for its output array. Note how the return value of the filter function, <code>_add_edge_filter</code>, is just 0! In our case, the output array contains all 0s, but we use the filter <em>exclusively for its side-effect</em>: adding an edge to the graph g when there is more than one unique value in the footprint. That‚Äôs cool.</p>
<p>Continuing, in this first RAG implementation, Vighnesh wanted to segment according to average color, so he further needed to iterate over each pixel/voxel/hypervoxel and keep a running total of the color and the pixel count. He used elements in the graph node dictionary for this and updated them using <code>ndindex</code>:</p>
<pre class="code literal-block"><span></span><span class="k">[code lang=python]</span>
<span class="na">for index in np.ndindex(labels.shape):</span>
    <span class="na">current</span> <span class="o">=</span> <span class="s">labels[index]</span>
<span class="s">    g.node[current]['pixel count'] += 1</span>
<span class="s">    g.node[current]['total color'] += image[index]</span>
<span class="k">[/code]</span>
</pre>


<p>Thus, together, numpy‚Äôs <code>nditer</code>, <code>ndindex</code>, and scipy.ndimage‚Äôs <code>generic_filter</code> provide a powerful way to perform a large variety of operations on n-dimensional arrays‚Ä¶ Much larger than I‚Äôd realised!</p>
<p>You can see Vighnesh‚Äôs complete pull request <a href="https://github.com/scikit-image/scikit-image/pull/1031">here</a> and follow his blog <a href="http://vcansimplify.wordpress.com/">here</a>.</p>
<p>If you use NumPy arrays and their massive bag of tricks, please cite the paper below!</p>
<p><span class="Z3988" title="ctx_ver=Z39.88-2004&amp;rft_val_fmt=info%3Aofi%2Ffmt%3Akev%3Amtx%3Ajournal&amp;rft.jtitle=Computing+in+Science+and+Engineering+13%2C+2+%282011%29+22-30&amp;rft_id=info%3Aarxiv%2F1102.1523v1&amp;rfr_id=info%3Asid%2Fresearchblogging.org&amp;rft.atitle=The+NumPy+array%3A+a+structure+for+efficient+numerical+computation&amp;rft.issn=&amp;rft.date=2011&amp;rft.volume=&amp;rft.issue=&amp;rft.spage=&amp;rft.epage=&amp;rft.artnum=&amp;rft.au=Stefan+Van+Der+Walt&amp;rft.au=S.+Chris+Colbert&amp;rft.au=Ga%C3%ABl+Varoquaux&amp;rfe_dat=bpr3.included=1;bpr3.tags=Computer+Science+%2F+Engineering%2CSoftware+Engineering">Stefan Van Der Walt, S. Chris Colbert, &amp; Ga√´l Varoquaux (2011). The NumPy array: a structure for efficient numerical computation <span style="font-style:italic;">Computing in Science and Engineering 13, 2 (2011) 22-30</span> arXiv: <a rev="review" href="http://arxiv.org/abs/1102.1523v1">1102.1523v1</a></span></p>
</div>
    </div>
    </article><article class="post h-entry post-text"><header><h1 class="post-title p-name"><a href="2014/06/18/elsevier-et-als-pricing-douchebaggery-exposed/" class="u-url">Elsevier et al's pricing douchebaggery exposed</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn">Juan Nunez-Iglesias</span></p>
            <p class="dateline"><a href="2014/06/18/elsevier-et-als-pricing-douchebaggery-exposed/" rel="bookmark"><time class="post-date published dt-published" datetime="2014-06-18T07:48:11+10:00" title="2014-06-18 07:48">2014-06-18 07:48</time></a></p>
                <p class="commentline">
        
    <a href="2014/06/18/elsevier-et-als-pricing-douchebaggery-exposed/#disqus_thread" data-disqus-identifier="cache/posts/2014/06/18/elsevier-et-als-pricing-douchebaggery-exposed.html">Comments</a>


        </p>
</div>
    </header><div class="e-content entry-content">
    <div>
<p></p>
<a href="http://www.econ.ucsb.edu/~tedb/">Ted Bergstrom</a> and a few colleagues have just come out with an <a href="http://www.pnas.org/content/early/2014/06/11/1403006111">epic paper</a> in which they reveal how much for-profit academic publishing companies charge university libraries, numbers that had previously been kept secret. The paper is ostensibly in the field of economics, but it could be more accurately described as "sticking-it-to-the-man-ology".
<p>This paragraph in the footnotes was particularly concise and fun to read:
</p>
<blockquote>Elsevier contested our contract request from Washington State University on the grounds that their pricing policy was a trade secret, and brought suit against the university. The Superior Court judge ruled that Washington State University could release the contracts to us. Elsevier and Springer also contested our request for contracts from the University of Texas (UT) System. The Texas state attorney general opined that the UT System was required to release copies of all of these contracts.</blockquote>
In other words: in the interest of full disclosure, <em>suck it, Elsevier!</em>
<p>The executive summary is that these companies will extort as much as they possibly can from individual universities, then do everything to keep that number a secret so that they can more freely extort <em>even more</em> from other universities. You can read more about it in <a href="http://news.sciencemag.org/economics/2014/06/how-much-did-your-university-pay-your-journals">Science</a>.</p>
<p>Oh, and in the ultimate twist of irony, the paper itself is behind PNAS's paywall. How much did your university pay for that?</p>
<p><span class="Z3988" title="ctx_ver=Z39.88-2004&amp;rft_val_fmt=info%3Aofi%2Ffmt%3Akev%3Amtx%3Ajournal&amp;rft.jtitle=Proceedings+of+the+National+Academy+of+Sciences&amp;rft_id=info%3Adoi%2F10.1073%2Fpnas.1403006111&amp;rfr_id=info%3Asid%2Fresearchblogging.org&amp;rft.atitle=Evaluating+big+deal+journal+bundles&amp;rft.issn=0027-8424&amp;rft.date=2014&amp;rft.volume=&amp;rft.issue=&amp;rft.spage=&amp;rft.epage=&amp;rft.artnum=http%3A%2F%2Fwww.pnas.org%2Fcgi%2Fdoi%2F10.1073%2Fpnas.1403006111&amp;rft.au=Bergstrom%2C+T.&amp;rft.au=Courant%2C+P.&amp;rft.au=McAfee%2C+R.&amp;rft.au=Williams%2C+M.&amp;rfe_dat=bpr3.included=1;bpr3.tags=Research+%2F+Scholarship%2COther%2CPublishing%2C+Policy%2C+Library+Science%2C+Funding%2C+Ethics">Bergstrom, T., Courant, P., McAfee, R., &amp; Williams, M. (2014). Evaluating big deal journal bundles <span style="font-style:italic;">Proceedings of the National Academy of Sciences</span> DOI: <a rev="review" href="http://dx.doi.org/10.1073/pnas.1403006111">10.1073/pnas.1403006111</a></span></p>
</div>
    </div>
    </article><article class="post h-entry post-text"><header><h1 class="post-title p-name"><a href="2014/03/15/an-update-on-mixing-java-and-python-with-fiji/" class="u-url">An update on mixing Java and Python with Fiji</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn">Juan Nunez-Iglesias</span></p>
            <p class="dateline"><a href="2014/03/15/an-update-on-mixing-java-and-python-with-fiji/" rel="bookmark"><time class="post-date published dt-published" datetime="2014-03-15T02:21:06+11:00" title="2014-03-15 02:21">2014-03-15 02:21</time></a></p>
                <p class="commentline">
        
    <a href="2014/03/15/an-update-on-mixing-java-and-python-with-fiji/#disqus_thread" data-disqus-identifier="cache/posts/2014/03/15/an-update-on-mixing-java-and-python-with-fiji.html">Comments</a>


        </p>
</div>
    </header><div class="e-content entry-content">
    <div>
<p></p>
<p>Two weeks ago I <a href="http://ilovesymposia.com/2014/02/26/fiji-jython/">posted</a> about invoking ImageJ functions from Python using Fiji‚Äôs Jython interpreter. A couple of updates on the topic: </p>
<p>First, I‚Äôve made a <a href="https://github.com/jni/fiji-python">repository</a> with a template project encapsulating my tips from that post. It‚Äôs very simple to get a Fiji Jython script working from that template. As an example, <a href="https://github.com/jni/snemi-eval">here‚Äôs</a> a script to evaluate segmentations using the metric used by the <a href="2014/03/15/an-update-on-mixing-java-and-python-with-fiji/brainiac2.mit.edu/SNEMI3D/">SNEMI3D segmentation challenge</a> (a slightly modified version of the adapted Rand error). </p>

<p>Second, this entire discussion might be rendered obsolete by two incredible projects from the <a href="http://www.cellprofiler.org/">CellProfiler</a> team: <a href="https://github.com/CellProfiler/python-javabridge">Python-Javabridge</a>, which allows Python to interact seamlessly with Java code, and <a href="https://github.com/CellProfiler/python-bioformats">Python-Bioformats</a>, which uses Python-Javabridge to read Bioformats images into Python. I have yet to play with them, but both look like cleaner alternatives to interact with ImageJ than my Jython scripting! At some point I‚Äôll write a post exploring these tools, but if you get to it before me, please mention it in the comments!</p>

<p></p>
</div>
    </div>
    </article>
</div>

        <nav class="postindexpager"><ul class="pager">
<li class="previous">
                <a href="index-5.html" rel="prev">Newer posts</a>
            </li>
            <li class="next">
                <a href="index-3.html" rel="next">Older posts</a>
            </li>
        </ul></nav><script>var disqus_shortname="nikolademo";(function(){var a=document.createElement("script");a.async=true;a.src="https://"+disqus_shortname+".disqus.com/count.js";(document.getElementsByTagName("head")[0]||document.getElementsByTagName("body")[0]).appendChild(a)}());</script><footer id="footer"><p>Contents ¬© 2019         <a href="mailto:jni.soma@fastmail.com">Juan Nunez-Iglesias</a> - Powered by         <a href="https://getnikola.com" rel="nofollow">Nikola</a>         
<a rel="license" href="https://creativecommons.org/licenses/by/4.0/">
<img alt="Creative Commons License BY" style="border-width:0; margin-bottom:12px;" src="https://i.creativecommons.org/l/by/4.0/88x31.png"></a></p>
            
        </footer>
</div>
    </div>
    <label for="sidebar-checkbox" class="sidebar-toggle"></label>
    
    
    
            <script src="assets/js/all-nocdn.js"></script><!-- fancy dates --><script>
    moment.locale("en");
    fancydates(0, "YYYY-MM-DD HH:mm");
    </script><!-- end fancy dates -->
</body>
</html>
