<!DOCTYPE html>
<html prefix="og: http://ogp.me/ns# article: http://ogp.me/ns/article# " lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Publications | I Love Symposia!</title>
<link href="../../assets/css/all-nocdn.css" rel="stylesheet" type="text/css">
<link href="https://fonts.googleapis.com/css?family=Playfair+Display:700,900" rel="stylesheet">
<meta name="theme-color" content="#5670d4">
<meta name="generator" content="Nikola (getnikola.com)">
<link rel="alternate" type="application/rss+xml" title="RSS" href="../../rss.xml">
<link rel="canonical" href="http://ilovesymposia.com/pages/publications/">
<!--[if lt IE 9]><script src="../../assets/js/html5.js"></script><![endif]--><meta name="author" content="Juan Nunez-Iglesias">
<meta property="og:site_name" content="I Love Symposia!">
<meta property="og:title" content="Publications">
<meta property="og:url" content="http://ilovesymposia.com/pages/publications/">
<meta property="og:description" content="In the work below, I've marked my own favourites with a dagger (†) and elaborated a bit about why I like them.
Takemura, S., Xu, C. S., et al. (2015). Synaptic circuits and their variations within dif">
<meta property="og:type" content="article">
<meta property="article:published_time" content="2014-04-30T03:49:35Z">
</head>
<body>
<a href="#content" class="sr-only sr-only-focusable">Skip to main content</a>

<!-- Header and menu bar -->
<div class="container">
      <header class="blog-header py-3"><div class="row nbb-header align-items-center">
          <div class="col-md-3 col-xs-2 col-sm-2" style="width: auto;">
            <button class="navbar-toggler navbar-light bg-light nbb-navbar-toggler" type="button" data-toggle="collapse" data-target=".bs-nav-collapsible" aria-controls="bs-navbar" aria-expanded="false" aria-label="Toggle navigation">
                <span class="navbar-toggler-icon"></span>
            </button>
            <div class="collapse bs-nav-collapsible bootblog4-search-form-holder">
                
            </div>
        </div>
          <div class="col-md-6 col-xs-10 col-sm-10 bootblog4-brand" style="width: auto;">
            <a class="navbar-brand blog-header-logo text-dark" href="http://ilovesymposia.com/">

            <span id="blog-title">I Love Symposia!</span>
        </a>
          </div>
            <div class="col-md-3 justify-content-end align-items-center bs-nav-collapsible collapse flex-collapse bootblog4-right-nav">
            <nav class="navbar navbar-light bg-white"><ul class="navbar-nav bootblog4-right-nav">
<li class="nav-item">
    <a href="index.md" id="sourcelink" class="nav-link">Source</a>
    </li>


                    
            </ul></nav>
</div>
    </div>
</header><nav class="navbar navbar-expand-md navbar-light bg-white static-top"><div class="collapse navbar-collapse bs-nav-collapsible" id="bs-navbar">
            <ul class="navbar-nav nav-fill d-flex w-100">
<li class="nav-item">
<a href="../../archive.html" class="nav-link">Archives</a>
                </li>
<li class="nav-item">
<a href="../../categories/index.html" class="nav-link">Tags</a>
                </li>
<li class="nav-item">
<a href="../../rss.xml" class="nav-link">RSS feed</a>

                
            </li>
</ul>
</div>
<!-- /.navbar-collapse -->
</nav>
</div>

<div class="container" id="content" role="main">
    <div class="body-content">
        <!--Body content-->
        
        
<article class="post-text storypage" itemscope="itemscope" itemtype="http://schema.org/Article"><header><h1 class="p-name entry-title" itemprop="headline name"><a href="." class="u-url">Publications</a></h1>

        

    </header><div class="e-content entry-content" itemprop="articleBody text">
    <div>
<p></p>
<p>In the work below, I've marked my own favourites with a dagger (†) and elaborated a bit about why I like them.</p>
<p>Takemura, S., Xu, C. S., <em>et al</em>. (2015). <a href="http://www.pnas.org/content/112/44/13711.full">Synaptic circuits and their variations within different columns in the visual system of Drosophila.</a> PNAS 112 (44), 13711-13716</p>
<p><strong>†</strong> Van der Walt, S., Schönberger, J. L.,<em><strong> </strong></em><strong>Nunez-Iglesias, J.</strong>, Boulogne F., Warner, J. D., Yager, N., Gouillart, E., Yu, T., and the scikit-image contributors (2014). <a href="https://peerj.com/articles/453/">scikit-image: image processing in Python.</a> PeerJ <span class="self-citation-volume">2</span>:<span class="self-citation-elocation">e453</span>, PeerJ PrePrints 2:e336v2</p>
<p></p>
<p style="text-align:left;padding-left:30px;"><em>Developing tools is not often the most glamorous part of science, but it can be among the most influential. Scikit-image is a fantastic image processing library that I have had the privilege of working on extensively. If you are taking part in the revolution in imaging in biology, read this to see if scikit-image can help you make sense of your data.</em></p>
<p class="p1"><strong>Nunez-Iglesias, J.</strong>, Kennedy, R., Plaza, S. M., Chakraborty, A., and Katz, W. T. (2014). <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3983515/">Graph-based active learning of agglomeration (GALA): a Python library to segment 2D and 3D neuroimages.</a> <i>Frontiers in neuroinformatics</i> 8, 34. doi:10.3389/fninf.2014.00034.</p>

<p class="p1"><strong>† Nunez-Iglesias, J.</strong>, Kennedy, R., Parag, T., Shi, J., and Chklovskii, D. B. (2013). <a href="http://www.plosone.org/article/info%3Adoi%2F10.1371%2Fjournal.pone.0071715">Machine Learning of Hierarchical Clustering to Segment 2D and 3D Images.</a> <i>PLoS ONE</i> 8, e71715. doi:10.1371/journal.pone.0071715.</p>

<p class="p1" style="padding-left:30px;"><em>I developed an algorithm, graph-based active learning of agglomeration (GALA), that can learn hierarchical agglomeration at all levels of the hierarchy, which is important for multi-scale image segmentation problems. It produced exceptional results in segmentation accuracy. In the <a href="http://brainiac2.mit.edu/SNEMI3D/home">SNEMI3D challenge</a>, our second submission was ranked 1st for 8 months in 2013. In Apr 2014 Neal Donnelly modified the <a href="https://github.com/janelia-flyem/gala">open source implementation</a> (detailed in the Frontiers in Neuroinformatics paper above) to regain first place, by a wide margin.</em></p>

<p class="p1">Hu, T., <strong>Nunez-Iglesias, J.</strong>, Vitaladevuni, S., Scheffer, L., Xu, S., Bolorizadeh, M., Hess, H., Fetter, R., and Chklovskii, D. (2013). Electron Microscopy Reconstruction of Brain Structure Using Sparse Representations over Learned Dictionaries. <i>IEEE transactions on medical imaging</i>. doi:10.1109/TMI.2013.2276018.</p>

<p class="p1">Grover, D.*, and <strong>Nunez-Iglesias, J.</strong>* (2012). <a href="http://online.liebertpub.com/doi/full/10.1089/cmb.2012.0036">Betamax: towards optimal sampling strategies for high-throughput screens.</a> <i>J. Comput. Biol.</i> 19, 776–784. doi:10.1089/cmb.2012.0036.</p>

<p class="p1">Glasner, D., Hu, T., <strong>Nunez-Iglesias, J.</strong>, Scheffer, L., Xu, S., Hess, H., Fetter, R., Chklovskii, D. B., and Basri, R. (2011). High resolution segmentation of neuronal tissues from low depth-resolution EM imagery. <i>EMMCVPR '11</i>, 1–12.</p>

<p class="p1"><strong>Nunez-Iglesias, J.</strong>, Liu, C.-C., Morgan, T. E., Finch, C. E., and Zhou, X. J. (2010). <a href="http://www.plosone.org/article/info%3Adoi%2F10.1371%2Fjournal.pone.0008898">Joint genome-wide profiling of miRNA and mRNA expression in Alzheimer's disease cortex reveals altered miRNA regulation.</a> <i>PLoS ONE</i> 5, e8898. doi:10.1371/journal.pone.0008898.</p>

<p class="p1">Mehan, M. R., <strong>Nunez-Iglesias, J.</strong>, Dai, C., Waterman, M. S., and Zhou, X. J. (2010). An integrative modular approach to systematically predict gene-phenotype associations. <i>BMC Bioinformatics</i> 11 Suppl 1, S62. doi:10.1186/1471-2105-11-S1-S62.</p>

<p class="p1">Mehan, M. R., <strong>Nunez-Iglesias, J.</strong>, Kalakrishnan, M., Waterman, M. S., and Zhou, X. J. (2009). An integrative network approach to map the transcriptome to the phenome. <i>J. Comput. Biol.</i> 16, 1023–1034. doi:10.1089/cmb.2009.0037.</p>

<p class="p1">Wang, W.*, <strong>Nunez-Iglesias, J.</strong>*, Luan, Y., and Sun, F. (2009). Usefulness and limitations of dK random graph models to predict interactions and functional homogeneity in biological networks under a pseudo-likelihood parameter estimation approach. <i>BMC Bioinformatics</i> 10, 277. doi:10.1186/1471-2105-10-277.</p>

<p class="p1">Xu, M., Kao, M.-C. J., <strong>Nunez-Iglesias, J.</strong>, Nevins, J. R., West, M., and Zhou, X. J. (2008). An integrative approach to characterize disease-specific pathways and their coordination: a case study in cancer. <i>BMC Genomics</i> 9 Suppl 1, S12. doi:10.1186/1471-2164-9-S1-S12.</p>

<p class="p1">Pan, F., Chiu, C.-H., Pulapura, S., Mehan, M. R., <strong>Nunez-Iglesias, J.</strong>, Zhang, K., Kamath, K., Waterman, M. S., Finch, C. E., and Zhou, X. J. (2007). Gene Aging Nexus: a web database and data mining platform for microarray data on aging. <i>Nucleic Acids Res</i> 35, D756–9. doi:10.1093/nar/gkl798.</p>

<p class="p1">Pan, F., Kamath, K., Zhang, K., Pulapura, S., Achar, A., <strong>Nunez-Iglesias, J.</strong>, Huang, Y., Yan, X., Han, J., Hu, H., et al. (2006). Integrative Array Analyzer: a software package for analysis of cross-platform and cross-species microarray data. <i>Bioinformatics</i> 22, 1665–1667. doi:10.1093/bioinformatics/btl163.</p>

<p class="p1">Noble, B., Abada, P., <strong>Nunez-Iglesias, J.</strong>, and Cannon, P. M. (2006). Recruitment of the adaptor protein 2 complex by the human immunodeficiency virus type 2 envelope protein is necessary for high levels of virus release. <i>J. Virol.</i> 80, 2924–2932. doi:10.1128/JVI.80.6.2924-2932.2006.</p>

<p class="p1">Zhou, X. J., Kao, M.-C. J., Huang, H., Wong, A., <strong>Nunez-Iglesias, J.</strong>, Primig, M., Aparicio, O. M., Finch, C. E., Morgan, T. E., and Wong, W. H. (2005). Functional annotation and network reconstruction through cross-platform integration of microarray data. <i>Nat. Biotechnol.</i> 23, 238–243. doi:10.1038/nbt1058.</p>

<p></p>
</div>
    </div>
    

</article><!--End of body content--><footer id="footer">
            Contents © 2019         <a href="mailto:jni.soma@gmail.com">Juan Nunez-Iglesias</a> - Powered by         <a href="https://getnikola.com" rel="nofollow">Nikola</a>         
            
        </footer>
</div>
</div>


        <script src="../../assets/js/all-nocdn.js"></script><script>
    baguetteBox.run('div#content', {
        ignoreClass: 'islink',
        captions: function(element) {
            return element.getElementsByTagName('img')[0].alt;
    }});
    </script>
</body>
</html>
